// @wf1/core-ui/esm2015/lib/full-width-container/full-width-container.component.js
import { Component, Input, TemplateRef } from "@angular/core";
import { TokenService } from "../services/token.service";
export class FullWidthContainerComponent {
    constructor(token) {
        this.token = token;
        this.isInitializing = true;
        this.isInitializing = token.getOauthToken() === null;
    }
    ngOnInit() {
        this.token.authTokenEmitter.subscribe(() => {
            this.isInitializing = false;
        });
    }
}
FullWidthContainerComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf1-full-width-container',
                template: "<!--\n  DEFAULT TEMPLATE DECLARATIONS FOR SUB-COMPONENTS\n-->\n\n<ng-template #defaultMainContent>\n  <p>\n    Lorem ipsum dolor sit amet, consectetur adipisicing elit. Aspernatur consectetur debitis, delectus deleniti dolore\n    earum explicabo facere facilis, fugiat itaque numquam provident quam quo quos sint veritatis vero voluptas voluptate?\n  </p>\n</ng-template>\n\n<!--\n  MAIN TEMPLATE DECLARATION\n-->\n<div id=\"core-interface-container\">\n  <wf1-application-header [titleText]=\"titleText\"></wf1-application-header>\n\n  <div class=\"wf1-loading-screen\" *ngIf=\"isInitializing\">\n    <h1>Initializing Application</h1>\n    <wf1-spinner></wf1-spinner>\n    <h3>Please wait while the application initializes.</h3>\n  </div>\n\n  <div class=\"main-content-container\" *ngIf=\"!isInitializing\">\n    <ng-container *ngTemplateOutlet=\"mainContentTemplate ? mainContentTemplate: defaultMainContent\"></ng-container>\n  </div>\n</div>\n"
            },] }
];
FullWidthContainerComponent.ctorParameters = () => [
    { type: TokenService }
];
FullWidthContainerComponent.propDecorators = {
    titleText: [{ type: Input }],
    mainContentTemplate: [{ type: Input }]
};
