// @wf1/core-ui/esm2015/lib/wfml-basemap-selector/wfml-basemap-selector.component.js
import { Component } from "@angular/core";
import { MapService } from "../services/map.service";
export class WFMLBasemapSelectorComponent {
    constructor(mapService) {
        this.mapService = mapService;
        this.basemaps = [];
    }
    ngOnInit() {
        if (this.mapService.isMapInitialized('map')) {
            this.initializeBasemaps();
        }
        else {
            this.mapService.mapInstanceEmitter.asObservable().subscribe(data => {
                this.initializeBasemaps();
            });
        }
    }
    initializeBasemaps() {
        this.basemaps = this.mapService.getBasemaps('map');
    }
    displayBasemap(index) {
        this.mapService.setBasemap('map', index);
    }
}
WFMLBasemapSelectorComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf1-wfml-basemap-selector',
                template: "<div class=\"basemap-list\">\n    <mat-card class=\"basemap\" *ngFor=\"let basemap of basemaps; let i = index\" (click)=\"displayBasemap(i)\">\n        <img alt=\"{{basemap.title}}\" mat-card-image src=\"assets/images/map-thumbnails/{{basemap.key}}.jpg\">\n        <mat-card-title>{{basemap.title}}</mat-card-title>\n    </mat-card>\n</div>\n",
                styles: [".basemap-list{display:flex}.basemap-list .basemap{flex:1 1 2rem;text-align:center;padding:0}.basemap-list .basemap .mat-card-image{width:100%;margin:0}.basemap-list .basemap .mat-card-title{padding:1.25rem;font-size:medium}"]
            },] }
];
WFMLBasemapSelectorComponent.ctorParameters = () => [
    { type: MapService }
];
