// @wf1/core-ui/esm2015/lib/search/components/filters/filters.component.js
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from "@angular/core";
import { FormBuilder } from "@angular/forms";
import { select, Store } from "@ngrx/store";
import * as momentInstance from "moment";
import { IncidentType } from "../../../application-icon/application-icon.model";
import { selectCurrentSearchParam } from "../../../utils";
import * as SearchActions from "../../store/actions";
const moment = momentInstance;
export class FiltersComponent {
    constructor(formBuilder, store) {
        this.formBuilder = formBuilder;
        this.store = store;
        this.filters = [];
        this.TOOLTIP_DELAY = 500;
        this.fromDateMin = [];
        this.toDateMax = [];
        this.DATE_FORMATS = {
            fullPickerInput: 'YYYY-MM-DD HH:mm',
            datePickerInput: 'YYYY-MM-DD',
            timePickerInput: 'HH:mm:ss',
            monthYearLabel: 'YYYY-MM',
            dateA11yLabel: 'YYYY-MMM-DD',
            monthYearA11yLabel: 'YYYY-MMM',
        };
        this.isShowingFilterOptions = new EventEmitter();
        this.activeFilters = {};
        this.showBody = false;
        this.lastTabPosition = -1;
        this.clearFilterIconType = IncidentType.NOT_SET;
        this.validateDate = (m) => {
            if (!m) { // empty value , so return valid format true
                return true;
            }
            else if (m.creationData) { //Moment
                let input = m.creationData().input;
                if (input.year) { //datepicker selection - input is a Date object - must be valid
                    return true;
                }
                else { //string input - check strict format using moment
                    return moment(input, this.DATE_FORMATS.datePickerInput, true).isValid() || moment(input, this.DATE_FORMATS.fullPickerInput, true).isValid();
                }
            }
            else {
                return true;
            }
        };
    }
    ngOnChanges(changes) {
        if (changes.filters) {
            this.createForm();
        }
    }
    ngOnInit() {
    }
    ngAfterViewInit() {
        this.getActiveFilters();
        setTimeout(() => {
            this.toggleActiveTabContent(true, true);
        }, 500);
    }
    createForm() {
        let fb = this.formBuilder;
        const config = this.filters.reduce(function (acc, f) {
            switch (f.type) {
                case 'dateRange':
                case 'dateTimeRange':
                    acc[f.param] = fb.array([fb.control(null), fb.control(null)]);
                    break;
                case 'single':
                    acc[f.param] = fb.control(null);
                    break;
                case 'multi':
                default:
                    acc[f.param] = fb.array(f.options.map(function () {
                        return fb.control(false);
                    }));
                    break;
            }
            return acc;
        }, {});
        this.filterForm = this.formBuilder.group(config);
    }
    checkboxUpdated(event) {
        const fv = this.filterForm.getRawValue();
        // console.log(fv)
        this.extractValuesFromForm(fv);
    }
    dateUpdated(event, param, fromTo) {
        let fromControl = this.getFormControlsFromArray(this.filterForm, param)[0];
        let toControl = this.getFormControlsFromArray(this.filterForm, param)[1];
        let hasError = false;
        let fromVal = fromTo == 'from' ? event.value : fromControl.value;
        let toVal = fromTo == 'to' ? event.value : toControl.value;
        if (fromTo == 'from' && !this.validateDate(fromVal)) {
            fromControl.setErrors({ 'incorrectFormat': true });
            hasError = true;
        }
        else if (fromTo == 'to' && !this.validateDate(toVal)) {
            toControl.setErrors({ 'incorrectFormat': true });
            hasError = true;
        }
        if (!hasError && fromVal && toVal) {
            fromControl.setErrors(null);
            toControl.setErrors(null);
            let toMoment = moment(toVal);
            let fromMoment = moment(fromVal);
            if (fromMoment.isAfter(toMoment)) {
                if (fromTo == 'from') {
                    fromControl.setErrors({ 'fromAfterTo': true });
                }
                else {
                    toControl.setErrors({ 'toBeforeFrom': true });
                }
                hasError = true;
            }
        }
        if (!hasError) {
            this.extractValuesFromForm(this.filterForm.getRawValue());
        }
    }
    getActiveFilters() {
        this.store.pipe(select(selectCurrentSearchParam(this.componentId, 'filters'))).subscribe(filters => {
            this.activeFilters = filters;
            setTimeout(() => {
                this.updateFormValues();
            }, 0);
        });
    }
    selectTab(tab) {
        // Toggle tab content if active tab is selected, or open tab content if tab clicked and currently tabs are hidden.
        this.toggleActiveTabContent(tab.isActive, this.showBody);
        if (tab.isActive || !this.showBody) {
            this.showBody = !this.showBody;
            this.isShowingFilterOptions.emit(this.showBody);
        }
        this.lastTabPosition = tab.position;
    }
    toggleActiveTabContent(isActiveTab, showBody) {
        let tabBodyWrapperEl = this.filterTabBodyEl.nativeElement.parentElement;
        if (isActiveTab) {
            if (showBody) {
                tabBodyWrapperEl.classList.remove('slidedown');
                tabBodyWrapperEl.classList.add('slideup');
            }
            else {
                tabBodyWrapperEl.classList.remove('slideup');
                tabBodyWrapperEl.classList.add('slidedown');
            }
        }
        else {
            tabBodyWrapperEl.classList.remove('slideup');
            tabBodyWrapperEl.classList.add('slidedown');
        }
    }
    clearFilter(param) {
        this.store.dispatch(new SearchActions.ClearFilterAction(param, this.componentId));
        let formArray = this.filterForm.get(param);
        for (let i = 0; i < formArray.length; i++) {
            let formControl = formArray.controls[i];
            formControl.reset();
        }
    }
    // Update from values from Redux
    updateFormValues() {
        let af = this.activeFilters;
        const formValue = this.filters.reduce(function (acc, f) {
            switch (f.type) {
                case 'dateRange':
                case 'dateTimeRange':
                    acc[f.param] = af && af[f.param] || [null, null];
                    break;
                case 'single':
                    acc[f.param] = ([].concat(af[f.param]))[0]; // ensure scalar, might be array
                    break;
                case 'multi':
                default:
                    acc[f.param] = f.options
                        .map(function (o) {
                        if (!af || !af[f.param])
                            return false;
                        return af[f.param].includes(o.value);
                    });
                    break;
            }
            return acc;
        }, {});
        this.filterForm.patchValue(formValue, { emitEvent: false });
    }
    // Convert form control boolean values to filter values
    extractValuesFromForm(filterValues) {
        const filter = this.filters.reduce(function (acc, f) {
            acc[f.param] = f;
            return acc;
        }, {});
        const validDate = (d) => {
            if (!d || !this.validateDate(d))
                return null;
            return d;
        };
        const activeFilters = Object.entries(filterValues).reduce(function (acc, [prop, val]) {
            let f = filter[prop];
            switch (f.type) {
                case 'dateRange':
                case 'dateTimeRange':
                    acc[prop] = [validDate(val[prop][0]), validDate(val[prop][1])];
                    break;
                case 'single':
                    acc[prop] = val;
                    break;
                case 'multi':
                default:
                    let vals = val
                        .map(function (v, i) { return v ? f.options[i].value : null; })
                        .filter(function (v) { return v; });
                    if (vals.length > 0)
                        acc[prop] = vals;
                    break;
            }
            return acc;
        }, {});
        if (this.filterForm.valid) {
            this.store.dispatch(new SearchActions.UpdateActiveFiltersAction(activeFilters, this.componentId));
        }
    }
    // Template functions
    getFormArrayNames(group) {
        return Object.keys(group.controls);
    }
    getFormControlsFromArray(group, prop) {
        const array = group.get(prop);
        return array.controls;
    }
    getSelectedFilterValues(param) {
        const filter = this.filters.find(function (f) { return f.param == param; });
        let af = this.activeFilters[param];
        switch (filter.type) {
            case 'dateRange':
            case 'dateTimeRange':
                let controls = this.getFormControlsFromArray(this.filterForm, param);
                if (controls) {
                    let format = filter.type == 'dateRange' ? this.DATE_FORMATS.datePickerInput : this.DATE_FORMATS.fullPickerInput;
                    let from = controls[0] && controls[0].value && this.validateDate(controls[0].value) ? moment(controls[0].value).format(format) : 'Any';
                    let to = controls[1] && controls[1].value && this.validateDate(controls[1].value) ? moment(controls[1].value).format(format) : 'Any';
                    return from == "Any" && to == "Any" ? "Any" : from + ' - ' + to;
                }
                else {
                    return "Any";
                }
            case 'single':
                let val = ([].concat(af))[0];
                if (!val)
                    return 'Any';
                return filter.options.find(function (o) { return o.value == val; }).label;
            case 'multi':
            default:
                if (!af || !af.length)
                    return 'Any';
                return filter.options.find(function (o) { return o.value == af[0]; }).label;
            // return (this.activeFilters
            //     && this.labelLookup
            //     && this.activeFilters[param]
            //     && this.activeFilters[param].length > 0
            //     //&& this.activeFilters[param][0]
            //     && this.labelLookup[param])
            //     ? this.labelLookup[param][this.activeFilters[param][0]]
            //     : "Any";
        }
    }
}
FiltersComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf1-filters',
                template: "<form class=\"wf1-filters\" [formGroup]=\"filterForm\">\n\t<mat-tab-group>\n\t\t<mat-tab #tab=\"matTab\" class=\"filter-tab\" *ngFor=\"let param of getFormArrayNames(filterForm); let i = index\">\n\t\t\t<ng-template matTabLabel>\n\t\t\t\t<div class=\"filter-tab__header\" (click)=\"selectTab(tab)\" (keydown)=\"selectTab(tab)\">\n\t\t\t\t\t<span class=\"filter-tab__header__title\">{{filters[i].label}}</span>\n\t\t\t\t\t<span class=\"filter-tab__header__subtitle\">\n\t\t\t\t\t\t{{ getSelectedFilterValues( param ) }}\n\t\t\t\t\t</span>\n\t\t\t\t</div>\n\t\t\t</ng-template>\n\t\t\t<ng-template matTabContent>\n\t\t\t\t<div class=\"filter-tab__body\" #filtertabbody>\n\t\t\t\t\t<div class=\"filter-tab__body__icon\">\n\t\t\t\t\t\t<wf1-application-icon class=\"filter-tab__body__clear\" iconType=\"filter-cancel\" [colourCode]=\"clearFilterIconType\" (click)=\"clearFilter(param)\" matTooltip=\"Clear Filter\" [matTooltipShowDelay]=\"TOOLTIP_DELAY\"></wf1-application-icon>\n\t\t\t\t\t</div>\n          \t\t\t<div class=\"filter-tab__body__options\">\n\t\t\t\t\t\t<ng-container *ngIf=\"filters[i].type == 'dateRange' || filters[i].type == 'dateTimeRange'\">\n\t\t\t\t\t\t\t<mat-form-field class=\"date-picker-input\">\n\t\t\t\t\t\t\t\t<input matInput [owlDateTime]=\"fromDT\" (dateTimeChange)=\"dateUpdated($event, param, 'from')\"\n\t\t\t\t\t\t\t\t\t   placeholder=\"From\"\n\t\t\t\t\t\t\t\t\t   [formControl]=\"getFormControlsFromArray(filterForm, param)[0]\"\n\t\t\t\t\t\t\t\t\t   wf1CoreDateMask>\n\t\t\t\t\t\t\t\t<button mat-button [owlDateTimeTrigger]=\"fromDT\">\n\t\t\t\t\t\t\t\t\t<mat-icon>event</mat-icon>\n\t\t\t\t\t\t\t\t</button>\n\t\t\t\t\t\t\t\t<owl-date-time #fromDT pickerType=\"{{filters[i].type == 'dateRange'? 'calendar': 'both'}}\"\n\t\t\t\t\t\t\t\t\t\t\t   [showSecondsTimer]=\"false\"></owl-date-time>\n\t\t\t\t\t\t\t\t<mat-error *ngIf=\"getFormControlsFromArray(filterForm, param)[0].hasError('incorrectFormat') || getFormControlsFromArray(filterForm, param)[0].hasError('owlDateTimeParse')\">Format must be {{filters[i].type == 'dateRange'? DATE_FORMATS.datePickerInput: DATE_FORMATS.fullPickerInput}}</mat-error>\n\t\t\t\t\t\t\t\t<mat-error *ngIf=\"getFormControlsFromArray(filterForm, param)[0].hasError('fromAfterTo')\">\"From\" date cannot be after \"To\" date.</mat-error>\n\t\t\t\t\t\t\t</mat-form-field>\n\t\t\t\t\t\t\t<mat-form-field class=\"date-picker-input\">\n\t\t\t\t\t\t\t\t<input matInput [owlDateTime]=\"toDT\" (dateTimeChange)=\"dateUpdated($event, param,'to')\"\n\t\t\t\t\t\t\t\t\t   placeholder=\"To\"\n\t\t\t\t\t\t\t\t\t   [formControl]=\"getFormControlsFromArray(filterForm, param)[1]\"\n\t\t\t\t\t\t\t\t\t   wf1CoreDateMask>\n\t\t\t\t\t\t\t\t<button mat-button [owlDateTimeTrigger]=\"toDT\">\n\t\t\t\t\t\t\t\t\t<mat-icon>event</mat-icon>\n\t\t\t\t\t\t\t\t</button>\n\t\t\t\t\t\t\t\t<owl-date-time #toDT pickerType=\"{{filters[i].type == 'dateRange'? 'calendar': 'both'}}\"\n\t\t\t\t\t\t\t\t\t\t\t   [showSecondsTimer]=\"false\" ></owl-date-time>\n\t\t\t\t\t\t\t\t<mat-error *ngIf=\"getFormControlsFromArray(filterForm, param)[1].hasError('incorrectFormat') || getFormControlsFromArray(filterForm, param)[1].hasError('owlDateTimeParse')\">Format must be {{filters[i].type == 'dateRange'? DATE_FORMATS.datePickerInput: DATE_FORMATS.fullPickerInput}}</mat-error>\n\t\t\t\t\t\t\t\t<mat-error *ngIf=\"getFormControlsFromArray(filterForm, param)[1].hasError('toBeforeFrom')\">\"To\" date cannot be before \"From\" date.</mat-error>\n\t\t\t\t\t\t\t</mat-form-field>\n\t\t\t\t\t\t</ng-container>\n\n                        <ng-container *ngIf=\"filters[i].type == 'single'\">\n\t\t\t\t\t\t\t<mat-radio-group  \n                                class=\"filter-tab__body__radiogroup\" \n                                [formControl]=\"filterForm.get( param )\" \n                                (change)=\"checkboxUpdated($event);\"\n                            >\n\t\t\t\t\t\t\t\t<mat-radio-button\n                                    *ngFor=\"let filterControl of filters[ i ].options; let j = index\" \n                                    class=\"filter-tab__body__radiogroup__radiobutton\" \n                                    [value]=\"filters[i].options[j].value\"\n                                >\n                                    <span class=\"filter-tab__body__radiogroup__radiobutton__label\">\n                                        {{filters[i].options[j].label}}\n                                    </span>\n\t\t\t\t\t\t\t\t</mat-radio-button>\n\t\t\t\t\t\t\t</mat-radio-group>\n\t\t\t\t\t\t</ng-container>\n\n                        <ng-container *ngIf=\"filters[i].type == 'multi' || !filters[i].type\">\n\t\t\t\t\t\t\t<ng-container *ngFor=\"let filterControl of getFormControlsFromArray(filterForm, param); let j = index\" >\n\t\t\t\t\t\t\t\t<mat-checkbox \n                                    class=\"filter-tab__body__checkbox\" \n                                    [formControl]=\"filterControl\" \n                                    *ngIf=\"filters[i].options[j]\" \n                                    (change)=\"checkboxUpdated($event);\"\n                                >\n\t\t\t\t\t\t\t\t    <span class=\"filter-tab__body__checkbox__label\">\n\t\t\t\t\t\t\t\t\t    {{filters[i].options[j].label}}\n\t\t\t\t\t\t\t\t    </span>\n\t\t\t\t\t\t\t\t</mat-checkbox>\n\t\t\t\t\t\t\t</ng-container>\n\t\t\t\t\t\t</ng-container>\n                    </div>\n\t\t\t\t</div>\n\t\t\t</ng-template>\n\t\t</mat-tab>\n\t</mat-tab-group>\n</form>\n",
                styles: ["::ng-deep .slideup{max-height:0!important;overflow-y:hidden!important;transition:max-height .8s ease-in!important}::ng-deep .slidedown{max-height:500px!important;overflow-y:hidden!important;transition:max-height .8s ease-out!important}"]
            },] }
];
FiltersComponent.ctorParameters = () => [
    { type: FormBuilder },
    { type: Store }
];
FiltersComponent.propDecorators = {
    filters: [{ type: Input }],
    componentId: [{ type: Input }],
    isShowingFilterOptions: [{ type: Output }],
    filterTabBodyEl: [{ type: ViewChild, args: ['filtertabbody',] }]
};
