// @wf1/core-ui/esm2015/lib/search/store/actions.js
export var SearchActionTypes;
(function (SearchActionTypes) {
    SearchActionTypes["UPDATE_SEARCH_QUERY"] = "[ Search ] Update search query text";
    SearchActionTypes["RESET_SEARCH_QUERY"] = "[ Search ] Reset search query text";
    SearchActionTypes["UPDATE_SORT"] = "[ Search ] Update active sort";
    SearchActionTypes["TOGGLE_SORT_MODAL"] = "[ Search ] Toggle sort modal visibility";
    SearchActionTypes["UPDATE_ACTIVE_FILTERS"] = "[ Search ] Update active filters";
    SearchActionTypes["UPDATE_HIDDEN_FILTERS"] = "[ Search ] Update hidden filters";
    SearchActionTypes["CLEAR_FILTER"] = "[ Search ] Clear filter";
    SearchActionTypes["CLEAR_ALL_FILTERS"] = "[ Search ] Clear all filters";
    SearchActionTypes["REFRESH_SEARCH"] = "[ Search ] Refresh search";
})(SearchActionTypes || (SearchActionTypes = {}));
export class UpdateSearchQueryAction {
    constructor(query, componentId) {
        this.query = query;
        this.componentId = componentId;
        this.type = SearchActionTypes.UPDATE_SEARCH_QUERY;
    }
}
export class ResetSearchQueryAction {
    constructor(componentId) {
        this.componentId = componentId;
        this.type = SearchActionTypes.RESET_SEARCH_QUERY;
    }
}
export class UpdateSortAction {
    constructor(param, direction, componentId) {
        this.param = param;
        this.direction = direction;
        this.componentId = componentId;
        this.type = SearchActionTypes.UPDATE_SORT;
    }
}
export class ToggleSortModalAction {
    constructor(componentId) {
        this.componentId = componentId;
        this.type = SearchActionTypes.TOGGLE_SORT_MODAL;
    }
}
export class UpdateActiveFiltersAction {
    constructor(filters, componentId) {
        this.filters = filters;
        this.componentId = componentId;
        this.type = SearchActionTypes.UPDATE_ACTIVE_FILTERS;
    }
}
export class UpdateHiddenFiltersAction {
    constructor(filters, componentId) {
        this.filters = filters;
        this.componentId = componentId;
        this.type = SearchActionTypes.UPDATE_HIDDEN_FILTERS;
    }
}
export class ClearFilterAction {
    constructor(filterToRemove, componentId) {
        this.filterToRemove = filterToRemove;
        this.componentId = componentId;
        this.type = SearchActionTypes.CLEAR_FILTER;
    }
}
export class ClearAllFiltersAction {
    constructor(componentId) {
        this.componentId = componentId;
        this.type = SearchActionTypes.CLEAR_ALL_FILTERS;
    }
}
export class RefreshSearchAction {
    constructor(componentId) {
        this.componentId = componentId;
        this.type = SearchActionTypes.REFRESH_SEARCH;
    }
}
