// @wf1/core-ui/esm2015/lib/search/components/search-bar/search-bar.component.js
import { Component, ContentChild, Input } from "@angular/core";
import { FormControl } from "@angular/forms";
// Components
import { SortComponent } from "../sort/sort.component";
// External
import { select, Store } from "@ngrx/store";
import { distinctUntilChanged } from "rxjs/operators";
import * as SearchActions from "../../store/actions";
import { IncidentType } from "../../../application-icon/application-icon.model";
import { selectCurrentSearchParam } from "../../../utils";
export class SearchBarComponent {
    constructor(store) {
        this.store = store;
        this.TOOLTIP_DELAY = 500;
        this.count = 0;
        this.showSortButton = false;
        this.searchControl = new FormControl();
        this.clearFilterIconType = IncidentType.NOT_SET;
        this.isResetSearch = false;
    }
    set sortModal(modal) {
        this.showSortButton = modal ? true : false;
    }
    ngOnInit() {
        this.resetSearch();
        this.listenSearchChanges();
    }
    resetSearch() {
        this.isResetSearch = true;
        this.store.dispatch(new SearchActions.ResetSearchQueryAction(this.componentId));
    }
    listenSearchChanges() {
        this.store.pipe(select(selectCurrentSearchParam(this.componentId, 'query'))).subscribe((searchTerm) => {
            // console.log("this.searchControl.setValue change: ", searchTerm);
            // console.log("this.searchControl.setValue isResetSearch: " + this.isResetSearch);
            if (this.isResetSearch && this.searchControl.value == searchTerm) {
                // console.log("isReset true and same value so won't trigger searchControl observable.  reset isReset to false;")
                this.isResetSearch = false;
            }
            this.searchControl.setValue(searchTerm, { emitEvent: true });
        });
        this.searchControl.valueChanges.pipe(distinctUntilChanged()).subscribe((searchTerm) => {
            // console.log("this.searchControl.valueChanges isResetSearch: " + this.isResetSearch);
            // console.log("searchControl value change: ", searchTerm);
            if (this.isResetSearch) {
                // console.log("this.searchControl.valueChanges reset to false");
                this.isResetSearch = false;
            }
            else {
                // console.log("this.searchControl.valueChanges update search");
                this.updateSearch(searchTerm);
            }
        });
    }
    refreshSearch() {
        this.store.dispatch(new SearchActions.RefreshSearchAction(this.componentId));
    }
    updateSearch(searchTerm) {
        this.store.dispatch(new SearchActions.UpdateSearchQueryAction(searchTerm, this.componentId));
    }
    toggleSortModal() {
        this.store.dispatch(new SearchActions.ToggleSortModalAction(this.componentId));
    }
    clearFilters() {
        this.store.dispatch(new SearchActions.ClearAllFiltersAction(this.componentId));
    }
}
SearchBarComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf1-search-bar',
                template: "<div class=\"side-panel-toolbar search\">\n\t<div *ngIf=\"showSortButton\" class=\"search__sort\" matTooltip=\"Sort\" [matTooltipShowDelay]=\"TOOLTIP_DELAY\" (click)=\"toggleSortModal()\" (keydown)=\"toggleSortModal()\">\n\t\t<mat-icon class=\"search__sort__icon\">sort</mat-icon>\n\t</div>\n\t<ng-content select=\"wf1-sort-modal\"></ng-content>\n\t<div class=\"search__count\">\n\t\t<span class=\"search__count__number\">{{count}}</span>\n\t</div>\n\t<div class=\"search__input\">\n\t\t<mat-form-field aria-label=\"Search...\" class=\"search__input__control\">\n\t\t\t<input matInput placeholder=\"Search...\" [formControl]=\"searchControl\">\n\t\t</mat-form-field>\n\t</div>\n\t<div class=\"search__cancel\" (click)=\"resetSearch()\" (keydown)=\"resetSearch()\" matTooltip=\"Clear Search\" [matTooltipShowDelay]=\"TOOLTIP_DELAY\">\n\t\t<mat-icon class=\"search__cancel__icon\">close</mat-icon>\n\t</div>\n\t<div class=\"search__refresh\" (click)=\"refreshSearch()\" (keydown)=\"refreshSearch()\" matTooltip=\"Refresh Search\" [matTooltipShowDelay]=\"TOOLTIP_DELAY\">\n\t\t<mat-icon class=\"search__refresh__icon\">refresh</mat-icon>\n\t</div>\n\t<div class=\"search__clear-filters\" (click)=\"clearFilters()\" (keydown)=\"clearFilters()\" matTooltip=\"Clear Filters\" [matTooltipShowDelay]=\"TOOLTIP_DELAY\">\n\t\t<wf1-application-icon iconType=\"filter-cancel\" [colourCode]=\"clearFilterIconType\"></wf1-application-icon>\n\t</div>\n</div>\n"
            },] }
];
SearchBarComponent.ctorParameters = () => [
    { type: Store }
];
SearchBarComponent.propDecorators = {
    count: [{ type: Input }],
    componentId: [{ type: Input }],
    sortModal: [{ type: ContentChild, args: [SortComponent,] }]
};
