// @wf1/core-ui/esm2015/lib/interceptors/authentication-interceptor.js
import { Injectable, Injector } from "@angular/core";
import { AppConfigService } from "../services/app-config.service";
import { TokenService } from "../services/token.service";
/** Pass untouched request through to the next request handler. */
export class AuthenticationInterceptor {
    constructor(injector) {
        this.injector = injector;
    }
    intercept(req, next) {
        // console.log(this.appConfig.getConfig());
        let processedRequest = req;
        if (this.isUrlSecured(req.url)) {
            this.token = this.injector.get(TokenService);
            this.authToken = this.token.getOauthToken();
            if (this.authToken) {
                processedRequest = req.clone({
                    headers: req.headers.set('Authorization', 'Bearer ' + this.authToken)
                });
            }
        }
        return next.handle(processedRequest);
    }
    isUrlSecured(url) {
        let isSecured = false;
        const config = this.injector.get(AppConfigService).getConfig();
        if (config && config.rest) {
            for (let endpoint in config.rest) {
                if (url.startsWith(config.rest[endpoint])) {
                    isSecured = true;
                    break;
                }
            }
        }
        return isSecured;
    }
}
AuthenticationInterceptor.decorators = [
    { type: Injectable }
];
AuthenticationInterceptor.ctorParameters = () => [
    { type: Injector }
];
