// @wf1/orgunit-rest-api/esm2015/api/orgUnitList.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Wildfire Org Unit REST API
 * The Wildfire Org Unit REST API.
 *
 * OpenAPI spec version: 1.0.0-SNAPSHOT
 *
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */
/* tslint:disable:no-unused-variable member-ordering */
import { Inject, Injectable, Optional } from '@angular/core';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { CustomHttpUrlEncodingCodec } from '../encoder';
import { BASE_PATH } from '../variables';
import { OrgUnitConfiguration } from '../configuration';
export class OrgUnitListService {
    /**
     * @param {?} httpClient
     * @param {?} basePath
     * @param {?} configuration
     */
    constructor(httpClient, basePath, configuration) {
        this.httpClient = httpClient;
        this.basePath = 'https://https://d1api.vividsolutions.com/wforg-org-unit-api';
        this.defaultHeaders = new HttpHeaders();
        this.configuration = new OrgUnitConfiguration();
        if (basePath) {
            this.basePath = basePath;
        }
        if (configuration) {
            this.configuration = configuration;
            this.basePath = basePath || configuration.basePath || this.basePath;
        }
    }
    /**
     * @param {?} consumes string[] mime-types
     * @return {?} true: consumes contains 'multipart/form-data', false: otherwise
     */
    canConsumeForm(consumes) {
        /** @type {?} */
        const form = 'multipart/form-data';
        for (let consume of consumes) {
            if (form === consume) {
                return true;
            }
        }
        return false;
    }
    /**
     * @param {?=} effectiveAsOfDate
     * @param {?=} orgUnitName
     * @param {?=} integerAlias
     * @param {?=} characterAlias
     * @param {?=} orgUnitTypeCode
     * @param {?=} orgUnitAreaTypeCode
     * @param {?=} point
     * @param {?=} pageNumber
     * @param {?=} pageRowCount
     * @param {?=} orderBy
     * @param {?=} restVersion
     * @param {?=} observe
     * @param {?=} reportProgress
     * @return {?}
     */
    getOrgUnitList1(effectiveAsOfDate, orgUnitName, integerAlias, characterAlias, orgUnitTypeCode, orgUnitAreaTypeCode, point, pageNumber, pageRowCount, orderBy, restVersion, observe = 'body', reportProgress = false) {
        /** @type {?} */
        let queryParameters = new HttpParams({ encoder: new CustomHttpUrlEncodingCodec() });
        if (effectiveAsOfDate !== undefined) {
            queryParameters = queryParameters.set('effectiveAsOfDate', /** @type {?} */ (effectiveAsOfDate));
        }
        if (orgUnitName !== undefined) {
            queryParameters = queryParameters.set('orgUnitName', /** @type {?} */ (orgUnitName));
        }
        if (integerAlias !== undefined) {
            queryParameters = queryParameters.set('integerAlias', /** @type {?} */ (integerAlias));
        }
        if (characterAlias !== undefined) {
            queryParameters = queryParameters.set('characterAlias', /** @type {?} */ (characterAlias));
        }
        if (orgUnitTypeCode !== undefined) {
            queryParameters = queryParameters.set('orgUnitTypeCode', /** @type {?} */ (orgUnitTypeCode));
        }
        if (orgUnitAreaTypeCode !== undefined) {
            queryParameters = queryParameters.set('orgUnitAreaTypeCode', /** @type {?} */ (orgUnitAreaTypeCode));
        }
        if (point !== undefined) {
            queryParameters = queryParameters.set('point', /** @type {?} */ (point));
        }
        if (pageNumber !== undefined) {
            queryParameters = queryParameters.set('pageNumber', /** @type {?} */ (pageNumber));
        }
        if (pageRowCount !== undefined) {
            queryParameters = queryParameters.set('pageRowCount', /** @type {?} */ (pageRowCount));
        }
        if (orderBy !== undefined) {
            queryParameters = queryParameters.set('orderBy', /** @type {?} */ (orderBy));
        }
        /** @type {?} */
        let headers = this.defaultHeaders;
        if (restVersion !== undefined && restVersion !== null) {
            headers = headers.set('Rest-Version', String(restVersion));
        }
        /** @type {?} */
        let httpHeaderAccepts = [
            'application/json',
            'application/xml'
        ];
        /** @type {?} */
        let httpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
        if (httpHeaderAcceptSelected != undefined) {
            headers = headers.set("Accept", httpHeaderAcceptSelected);
        }
        /** @type {?} */
        let consumes = [];
        return this.httpClient.get(`${this.basePath}/orgUnits`, {
            params: queryParameters,
            withCredentials: this.configuration.withCredentials,
            headers: headers,
            observe: observe,
            reportProgress: reportProgress
        });
    }
}
OrgUnitListService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OrgUnitListService.ctorParameters = () => [
    { type: HttpClient },
    { type: String, decorators: [{ type: Optional }, { type: Inject, args: [BASE_PATH,] }] },
    { type: OrgUnitConfiguration, decorators: [{ type: Optional }] }
];
if (false) {
    /** @type {?} */
    OrgUnitListService.prototype.basePath;
    /** @type {?} */
    OrgUnitListService.prototype.defaultHeaders;
    /** @type {?} */
    OrgUnitListService.prototype.configuration;
    /** @type {?} */
    OrgUnitListService.prototype.httpClient;
}
