// @wf1/core-ui/esm2015/lib/public-application-header.module.js
import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";
import { PublicAppHeaderActionItem, PublicAppHeaderMenuItem, PublicApplicationHeaderComponent } from "./public-application-header/public-application-header.component";
import { MatMenuModule } from "@angular/material/menu";
import { MatToolbarModule } from "@angular/material/toolbar";
import { MatIconModule } from "@angular/material/icon";
import { MatDividerModule } from "@angular/material/divider";
export { PublicAppHeaderActionItem, PublicAppHeaderMenuItem };
export class PublicApplicationHeaderModule {
}
PublicApplicationHeaderModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    MatIconModule,
                    MatMenuModule,
                    MatToolbarModule,
                    MatDividerModule,
                ],
                declarations: [
                    PublicApplicationHeaderComponent
                ],
                exports: [
                    PublicApplicationHeaderComponent
                ]
            },] }
];
