// @wf1/core-ui/esm2015/lib/services/util/auth-guard.js
import { Injectable } from "@angular/core";
import { Router } from "@angular/router";
import { TokenService } from "../token.service";
import { of } from "rxjs";
import * as i0 from "@angular/core";
import * as i1 from "../token.service";
import * as i2 from "@angular/router";
export class AuthGuard {
    constructor(tokenService, router) {
        this.tokenService = tokenService;
        this.router = router;
        this.baseScopes = []; //["WFRM.GET_CODE_TABLES","WFRM.GET_TOPLEVEL"];
        // console.log("using auth guard");
        this.tokenService.credentialsEmitter.subscribe(credentials => {
            this.credentials = credentials;
        });
    }
    canActivate(next, state) {
        let url = state.url;
        let result = this.checkLogin(url, this.baseScopes);
        return of(result);
    }
    checkLogin(url, scopes) {
        let isAuthorized = (this.credentials) ? this.tokenService.doesUserHaveApplicationPermissions(scopes) : false;
        if (!isAuthorized) {
            this.redirectToErrorPage();
        }
        return isAuthorized;
    }
    redirectToErrorPage() {
        // Navigate to the unauthorized page
        this.router.navigate([{ outlets: { root: ['unauthorized'] } }]);
    }
}
AuthGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function AuthGuard_Factory() { return new AuthGuard(i0.ɵɵinject(i1.TokenService), i0.ɵɵinject(i2.Router)); }, token: AuthGuard, providedIn: "root" });
AuthGuard.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
AuthGuard.ctorParameters = () => [
    { type: TokenService },
    { type: Router }
];
