// ng-pick-datetime-moment/__ivy_ngcc__/moment-adapter/moment-date-time.module.js
import { NgModule } from '@angular/core';
import { MomentDateTimeAdapter, OWL_MOMENT_DATE_TIME_ADAPTER_OPTIONS } from './moment-date-time-adapter.class';
import { OWL_MOMENT_DATE_TIME_FORMATS } from './moment-date-time-format.class';
import { DateTimeAdapter, OWL_DATE_TIME_FORMATS, OWL_DATE_TIME_LOCALE } from 'ng-pick-datetime';
import * as ɵngcc0 from '@angular/core';
var MomentDateTimeModule = (function () {
    function MomentDateTimeModule() {
    }
MomentDateTimeModule.ɵfac = function MomentDateTimeModule_Factory(t) { return new (t || MomentDateTimeModule)(); };
MomentDateTimeModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: MomentDateTimeModule });
MomentDateTimeModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: DateTimeAdapter,
            useClass: MomentDateTimeAdapter,
            deps: [OWL_DATE_TIME_LOCALE, OWL_MOMENT_DATE_TIME_ADAPTER_OPTIONS]
        },
    ] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MomentDateTimeModule, [{
        type: NgModule,
        args: [{
                providers: [
                    {
                        provide: DateTimeAdapter,
                        useClass: MomentDateTimeAdapter,
                        deps: [OWL_DATE_TIME_LOCALE, OWL_MOMENT_DATE_TIME_ADAPTER_OPTIONS]
                    },
                ]
            }]
    }], function () { return []; }, null); })();
    return MomentDateTimeModule;
}());
export { MomentDateTimeModule };
var ɵ0 = OWL_MOMENT_DATE_TIME_FORMATS;
var OwlMomentDateTimeModule = (function () {
    function OwlMomentDateTimeModule() {
    }
OwlMomentDateTimeModule.ɵfac = function OwlMomentDateTimeModule_Factory(t) { return new (t || OwlMomentDateTimeModule)(); };
OwlMomentDateTimeModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: OwlMomentDateTimeModule });
OwlMomentDateTimeModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [{ provide: OWL_DATE_TIME_FORMATS, useValue: ɵ0 }], imports: [MomentDateTimeModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OwlMomentDateTimeModule, [{
        type: NgModule,
        args: [{
                imports: [MomentDateTimeModule],
                providers: [{ provide: OWL_DATE_TIME_FORMATS, useValue: ɵ0 }]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(OwlMomentDateTimeModule, { imports: [MomentDateTimeModule] }); })();
    return OwlMomentDateTimeModule;
}());
export { OwlMomentDateTimeModule };
export { ɵ0 };
