// @wf1/core-ui/esm2015/lib/services/app-config.service.js
import { Injectable } from "@angular/core";
import { HttpBackend, HttpClient } from "@angular/common/http";
import { AsyncSubject } from "rxjs";
import { LibraryConfig } from "../config/library-config";
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../config/library-config";
export class AppConfigService {
    constructor(httpHandler, libConfig) {
        this.httpHandler = httpHandler;
        this.libConfig = libConfig;
        this.config = new AsyncSubject();
        this.configEmitter = this.config.asObservable();
        // console.log("initing app config service");
    }
    loadAppConfig() {
        // console.log("loading app config");
        let http = new HttpClient(this.httpHandler);
        return http.get(this.libConfig.configurationPath).toPromise().then((data) => {
            this.appConfig = data;
            this.config.next(this.appConfig);
            this.config.complete();
        });
    }
    getConfig() {
        return this.appConfig;
    }
}
AppConfigService.ɵprov = i0.ɵɵdefineInjectable({ factory: function AppConfigService_Factory() { return new AppConfigService(i0.ɵɵinject(i1.HttpBackend), i0.ɵɵinject(i2.LibraryConfig)); }, token: AppConfigService, providedIn: "root" });
AppConfigService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
AppConfigService.ctorParameters = () => [
    { type: HttpBackend },
    { type: LibraryConfig }
];
