// @wf1/core-ui/esm2015/lib/services/map.service.js
import { HostListener, Injectable } from "@angular/core";
import { Subject } from "rxjs";
import { AppConfigService } from "./app-config.service";
import { DatePipe } from "@angular/common";
import * as i0 from "@angular/core";
import * as i1 from "./app-config.service";
export class MapService {
    constructor(appConfig) {
        this.appConfig = appConfig;
        this.mapInstances = [];
        this.mapInstanceEmitter = new Subject();
        // WFML.BASE_URL = appConfig.getConfig().application.baseUrl + 'assets/js/wfml-v0.3.0/v0.3.0/wfml';
    }
    findMapInstance(mapName) {
        return this.mapInstances.find(instance => instance.name === mapName);
    }
    onResize() {
        // for (const mapInstance of this.mapInstances) {
        //   if (mapInstance.isInitialized && mapInstance.api) {
        //     setTimeout(() => mapInstance.api.map.resize(), 400);
        //   }
        // }
    }
    /**
     * Returns a boolean flag that indicates whether the target map instance has been initialized.
     * @param mapName Id value of the leaflet element on the page.
     */
    isMapInitialized(mapName) {
        const targetInstance = this.findMapInstance(mapName);
        return (targetInstance && targetInstance.isInitialized);
    }
    /**
     * initialize map
     * @param mapName string
     */
    initializeMap(mapName) {
        // let targetInstance = this.findMapInstance(mapName);
        // if (!targetInstance) {
        //   targetInstance = {
        //     name: mapName,
        //     isInitialized: false
        //   };
        //   this.mapInstances.push(targetInstance);
        // }
        // if (!targetInstance.isInitialized) {
        //   const config: ApplicationConfig = this.appConfig.getConfig();
        //   // console.log(`Initializing map: ${targetInstance.name}`);
        //   targetInstance.api = WFML.createMap('map', config.mapServiceConfig.map, config.mapServiceConfig.layerSettings);
        //   targetInstance.isInitialized = true;
        //   this.mapInstanceEmitter.next(mapName);
        // }
    }
    /**
     * get layer tree view
     * @param mapName string
     */
    getLayerTreeView(mapName) {
        const targetInstance = this.findMapInstance(mapName);
        return (targetInstance && targetInstance.isInitialized)
            ? targetInstance.api.layer.treeView()
            : null;
    }
    /**
     * set map layer visibility
     * @param mapName string
     * @param id layer id
     * @param isVisible visibility boolean
     */
    setLayerVisibility(mapName, id, isVisible) {
        const targetInstance = this.findMapInstance(mapName);
        if (targetInstance && targetInstance.isInitialized) {
            targetInstance.api.layer.visible(id, isVisible);
        }
    }
    /**
     * zoomToPoint zooms the map to a specific point.
     * @param mapName string
     * @param coordinates coordinates
     * @param zoomLevel number
     */
    zoomToPoint(mapName, coordinates, zoomLevel) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.navigation.zoom(coordinates, zoomLevel);
    }
    /**
     * stop edit
     * @param mapName string
     */
    stopEdit(mapName) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.edit.stop();
    }
    /**
     * hide all map layer objects
     * @param mapName string
     * @param type ued or string
     */
    hideAll(mapName, type) {
        const targetInstance = this.findMapInstance(mapName);
        const data = targetInstance.api.layer.treeView().child.find((e) => e.title === type).child;
        data.map((layer) => {
            layer.lyr.isVisible = false;
        });
    }
    /**
     * redray Layer
     * @param mapName string
     * @param layerName string
     */
    redrawLayer(mapName, layerName) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.layer.redraw(layerName);
    }
    /**
     * add Highlight
     * @param mapName string
     * @param lonlat longitude latatude array
     */
    addHighlight(mapName, lonlat) {
        const targetInstance = this.findMapInstance(mapName);
        this.clearHighlight(mapName);
        targetInstance.api.highlight.add(lonlat);
    }
    /**
     * clear all highlights on map.
     * @param mapName string
     */
    clearHighlight(mapName) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.highlight.clear();
    }
    /**
     * Load Report of Fires as markers from report of fires resources
     * @param mapName string
     * @param reportOfFires collection of report of fire resources from WFIM api
     * @param markerClickCallback callback function for markers
     */
    loadRoFMarkers(mapName, reportOfFires, markerClickCallback) {
        const targetInstance = this.findMapInstance(mapName);
        const markers = reportOfFires.map(rof => {
            const date = (rof.messageStatusTimestamp) ? new DatePipe('en-US').transform(new Date(rof.messageStatusTimestamp), 'yyyy-MM-dd HH:mm:ss') : '';
            let phoneNumber = '';
            if (rof.callerTelephone) {
                const numberSegments = rof.callerTelephone.replace(/\D/g, '').match(/(\d{3})(\d{3})(\d{4})/);
                if (numberSegments && numberSegments.length === 4) {
                    phoneNumber = `(${numberSegments[1]}) ${numberSegments[2]}-${numberSegments[3]}`;
                }
                else {
                    console.error(`RoF ${rof.wildfireYear}-${rof.reportOfFireNumber} phone number value '${rof.callerTelephone}' is not valid.`);
                }
            }
            return {
                "messageTypeCode": "Public Report of Fire",
                "messageStatusCode": rof.messageStatusCode,
                "reportOfFireNumber": rof.reportOfFireNumber,
                "reportOfFireLabel": rof.reportOfFireLabel,
                "wildfireYear": rof.wildfireYear,
                "hoverTitle": rof.publicReportTypeCode,
                "publicReportTypeCode": rof.publicReportTypeCode,
                "latitude": rof.latitude,
                "longitude": rof.longitude,
                "yearNumber": rof.reportOfFireLabel,
                "latLon": this.formatCoordinates([rof.longitude, rof.latitude]),
                "lastUpdated": date,
                "updatedBy": rof.messageStatusUserId,
                "valuesBeingThreatenedNote": rof.valuesBeingThreatenedNote,
                "fireSizeComparisonCode": rof.fireSizeComparisionCode,
                "callerName": rof.callerName,
                "callerTelephone": phoneNumber
            };
        });
        targetInstance.api.layer.visible(WFML.RoF.LAYER_ID, true);
        WFML.RoF.Util.clear(targetInstance.api);
        WFML.RoF.Util.addClusteredMarkers(targetInstance.api, markers, markerClickCallback);
    }
    /**
     * Load Incidents as markers from incident resources
     * @param mapName string
     * @param optionsCodeTables Code[]
     * @param orgOptionsCodeTables Code[]
     * @param incidents collection of report of fire resources from WFIM api
     * @param markerClickCallback callback function for markers
     */
    loadIncidentMarkers(mapName, incidents, optionsCodeTables, orgOptionsCodeTables, markerClickCallback) {
        const targetInstance = this.findMapInstance(mapName);
        const markers = incidents.map(incident => {
            const interfaceFireInd = (incident.incidentSituation) ? incident.incidentSituation.interfaceFireInd : false;
            const interfaceText = (interfaceFireInd) ? 'Interface' : 'Non Interface';
            const markerLabel = (incident.incidentLabel === `${incident.wildfireYear}-${incident.incidentNumberSequence}`) ? `${incident.incidentNumberSequence}` : incident.incidentLabel;
            const incidentTypeCode = optionsCodeTables['INCIDENT_TYPE_CODE'].find(option => option.code === incident.incidentTypeCode);
            const stageOfControlCode = (incident.incidentSituation) ? optionsCodeTables['STAGE_OF_CONTROL_CODE'].find(option => option.code === incident.incidentSituation.stageOfControlCode) : undefined;
            const fireCentreOrgUnitIdentifier = orgOptionsCodeTables['FIRE_CENTRE_CODE'].find(option => option.code === incident.fireCentreOrgUnitIdentifier);
            const zoneOrgUnitIdentifier = orgOptionsCodeTables['ZONE_CODE'].find(option => option.code === incident.zoneOrgUnitIdentifier);
            return {
                "wildfireYear": incident.wildfireYear,
                "incidentNumberSequence": incident.incidentNumberSequence,
                "hoverTitle": interfaceText,
                "incidentId": incident.incidentId,
                "incidentName": incident.incidentName,
                "incidentLocation": {
                    "latitude": incident.incidentLocation ? incident.incidentLocation.latitude : '',
                    "longitude": incident.incidentLocation ? incident.incidentLocation.longitude : ''
                },
                "incidentSituation": {
                    "interfaceFireInd": interfaceFireInd
                },
                "incidentStatusCode": incident.incidentStatusCode,
                "markerLabel": markerLabel,
                "incidentLabel": incident.incidentLabel,
                "probabilityOfInitialAttackSuccessCode": incident.probabilityOfInitialAttackSuccessCode,
                "suspectedCauseCategoryCode": incident.suspectedCauseCategoryCode,
                "incidentCategoryCode": incident.incidentCategoryCode,
                "incidentTypeCode": incidentTypeCode ? incidentTypeCode.description : '',
                "stageOfControlCode": stageOfControlCode ? stageOfControlCode.description : '',
                "fireCentreOrgUnitIdentifier": fireCentreOrgUnitIdentifier ? fireCentreOrgUnitIdentifier.description : '',
                "zoneOrgUnitIdentifier": zoneOrgUnitIdentifier ? zoneOrgUnitIdentifier.description : '',
                "latLon": incident.incidentLocation ? WFML.Location.format([incident.incidentLocation.longitude, incident.incidentLocation.latitude]) : [],
                "geographicDescription": incident.incidentLocation ? incident.incidentLocation.geographicDescription : '',
                "fireSizeHectares": incident.incidentSituation ? incident.incidentSituation.fireSizeHectares : ''
            };
        });
        targetInstance.api.layer.visible(WFML.Incident.LAYER_ID, true);
        WFML.Incident.Util.clear(targetInstance.api);
        WFML.Incident.Util.addClusteredMarkers(targetInstance.api, markers, markerClickCallback);
    }
    /**
     * Clear markers from the specified mapName and markerLayerId
     * @param mapName string
     * @param markerLayerId identifier for the marker layer tracked by the marker manager
     */
    clearMarkers(mapName, markerLayerId) {
        const targetInstance = this.findMapInstance(mapName);
        return targetInstance.api.marker.removeMarkers(markerLayerId);
    }
    /**
     * Set markers layer visibility from the specified mapName and markerLayerId
     * @param mapName string
     * @param markerLayerId identifier for the marker layer tracked by the marker manager
     * @param visible boolean for setting the visibility of the marker layer
     */
    showMarkers(mapName, markerLayerId, visible) {
        const targetInstance = this.findMapInstance(mapName);
        return targetInstance.api.marker.showMarkers(markerLayerId, visible);
    }
    /**
    * Add legend layers
    * @param mapName name of the map instance
    * @param layerId id of the target map layer the legend attaches to
    * @param legend structure based on WFML api for legends - refer to WFML test bed for examples
    */
    addLegend(mapName, layerId, legend) {
        const targetInstance = this.findMapInstance(mapName);
        if (targetInstance) {
            targetInstance.api.layer.legend(layerId, legend);
        }
    }
    /**
     * Clear selected point (pin) from the specified map instance
     */
    clearSelectPoint(mapName) {
        const targetInstance = this.findMapInstance(mapName);
        if (targetInstance) {
            targetInstance.api.selectPoint.clear();
        }
    }
    setMaxSearchDistance(mapName, withinDistance) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.search.setMaximumDistance(withinDistance);
    }
    findPlace(mapName, searchText) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.search.findPlace(searchText);
    }
    findRoad(mapName, searchText) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.search.findRoad(searchText);
    }
    findIntersection(mapName, roadName, intersectingRoadName) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.search.findIntersection(roadName, intersectingRoadName);
    }
    clearPlaceNameSearch(mapName) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.search.clear();
    }
    panToPlaceNameSearchLocation(mapName) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.search.panToAnchor();
    }
    zoomToPlaceNameSearchLocation(mapName) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.search.zoomToAnchor();
    }
    showCandidate(mapName, location) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.search.showCandidate(location);
    }
    clearCandidate(mapName) {
        const targetInstance = this.findMapInstance(mapName);
        // showCandidate() with empty params clears the candidate selection.
        targetInstance.api.search.showCandidate();
    }
    setAnchor(mapName, point) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.search.setAnchor(point);
    }
    setAnchorFromMap(mapName) {
        const targetInstance = this.findMapInstance(mapName);
        const point = targetInstance.api.selectPoint.get();
        if (point) {
            targetInstance.api.search.setAnchor(point);
        }
        return point;
    }
    getBasemaps(mapName) {
        const targetInstance = this.findMapInstance(mapName);
        return targetInstance.api.basemap.all();
    }
    setBasemap(mapName, baseMapIndex) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.basemap.switch(baseMapIndex);
    }
    registerPlaceNameSearchListener(mapName, callback) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.search.setResultHandler(callback);
    }
    registerMapClickListener(mapName, callback) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.callback.click(callback);
    }
    registerMapHoverListener(mapName, callback, hoverDelay) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.callback.hover(callback, hoverDelay);
    }
    registerMapSelectPointListener(mapName, callback) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.callback.selectPoint(callback);
    }
    clearPlaceNameSearchListener(mapName) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.search.setResultHandler();
    }
    clearMapClickListener(mapName) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.callback.click();
    }
    clearMapHoverListener(mapName) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.callback.hover();
    }
    clearMapSelectPointListener(mapName) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.callback.selectPoint();
    }
    formatCoordinates(coordinates) {
        return WFML.Location.format(coordinates);
    }
    parseCoordinates(url) {
        return WFML.Location.parse(url);
    }
    activateTool(mapName, toolName) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.tool.activate(toolName);
    }
    persistState(mapName, key, opt) {
        const targetInstance = this.findMapInstance(mapName);
        targetInstance.api.state.persist(key, opt);
    }
}
MapService.ɵprov = i0.ɵɵdefineInjectable({ factory: function MapService_Factory() { return new MapService(i0.ɵɵinject(i1.AppConfigService)); }, token: MapService, providedIn: "root" });
MapService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
MapService.ctorParameters = () => [
    { type: AppConfigService }
];
MapService.propDecorators = {
    onResize: [{ type: HostListener, args: ['resize',] }]
};
