// @wf1/core-ui/esm2015/lib/services/navigation.service.js
import { Injectable } from '@angular/core';
import { Location } from '@angular/common';
import { Router, NavigationEnd } from '@angular/router';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@angular/common";
export class NavigationService {
    constructor(router, location) {
        this.router = router;
        this.location = location;
        this.navigations = 0;
        this.router.events.subscribe((event) => {
            if (event instanceof NavigationEnd) {
                this.navigations += 1;
                // console.log( this.navigations, event.urlAfterRedirects )
            }
        });
    }
    back() {
        if (this.navigations > 1) {
            this.location.back();
        }
        else {
            this.router.navigateByUrl('/');
        }
    }
}
NavigationService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NavigationService_Factory() { return new NavigationService(i0.ɵɵinject(i1.Router), i0.ɵɵinject(i2.Location)); }, token: NavigationService, providedIn: "root" });
NavigationService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
NavigationService.ctorParameters = () => [
    { type: Router },
    { type: Location }
];
