// @wf1/core-ui/esm2015/lib/services/window-messaging.service.js
import { Injectable } from "@angular/core";
import { UUID } from "angular2-uuid";
import { Subject } from "rxjs";
import { MAPTOOL_WINDOW_NAME, MessageType } from "./window-messaging.state";
import * as i0 from "@angular/core";
export class WindowMessagingService {
    constructor() {
        this.children = [];
        if (window.opener) {
            this.parentWindow = {
                id: UUID.UUID(),
                isParent: true,
                window: window.opener,
                messageQueue: []
            };
            window.addEventListener('onbeforeunload', () => this.broadcastMessage(this.getParentWindowId(), { type: MessageType.WINDOW_CLOSING }));
        }
        else {
            window.name = MAPTOOL_WINDOW_NAME;
        }
        this.stream = new Subject();
        window.addEventListener('message', (message) => this.receiveMessage(message.data));
        // setInterval( () => {
        //   if (this.parentWindow) {
        //
        //   } else {
        //     for (let currentChild of this.children) {
        //
        //     }
        //   }
        // }, 1000);
    }
    getParentWindowId() {
        return this.parentWindow ? this.parentWindow.id : null;
    }
    getWindowId(windowName) {
        const targetWindowIndex = this.children.findIndex(currentChild => currentChild.windowName === windowName);
        let targetWindow = (targetWindowIndex > -1) ? this.children[targetWindowIndex] : null;
        if (targetWindow && targetWindow.window && targetWindow.window.closed) {
            // If the window was closed, remove the object from the list of child windows.
            this.children.splice(targetWindowIndex, 1);
            targetWindow = null;
        }
        return (targetWindow) ? targetWindow.id : null;
    }
    broadcastAction(windowId, action) {
        this.broadcastMessage(windowId, {
            type: MessageType.ACTION,
            action
        });
    }
    openWindow(url, windowName = 'default', windowConfig = {
        fullscreen: true,
        rememberWindowSize: true
    }) {
        const windowFeatures = this.getWindowFeaturesSettings(windowConfig);
        const newWindow = {
            id: UUID.UUID(),
            isParent: false,
            url,
            windowName,
            windowConfig,
            window: window.open(url, windowName, windowFeatures),
            messageQueue: []
        };
        this.children.push(newWindow);
        return newWindow.id;
    }
    subscribeToMessageStream(listenerFunction) {
        this.stream.subscribe(message => listenerFunction(message));
        if (this.getParentWindowId() !== null) {
            // Once a listener has been registered, tell the parent to send along any queued messages
            this.broadcastMessage(this.getParentWindowId(), { type: MessageType.READY });
        }
    }
    focusWindow(windowId) {
        let targetWindow;
        if (windowId === this.getParentWindowId()) { // Children will only message their parent
            // This window.open call doesn't actually open a window, but is necessary for window.focus() to work on the parent window
            targetWindow = window.open('', MAPTOOL_WINDOW_NAME);
        }
        else {
            const target = this.getWindow(windowId);
            if (target && target.window) {
                targetWindow = target.window;
            }
        }
        targetWindow.focus();
    }
    broadcastAcknowledge(packet) {
        let target;
        if (this.getParentWindowId()) { // Children will only message their parent
            target = this.parentWindow;
        }
        else {
            target = this.children.find(child => child.windowName === packet.originWindowName);
        }
        if (target) {
            const messagePacket = {
                id: packet.id,
                originWindowName: window.name,
                date: new Date(),
                message: { type: MessageType.ACKNOWLEDGE }
            };
            target.window.postMessage(messagePacket, '*');
        }
    }
    broadcastMessage(windowId, message) {
        let target;
        if (this.getParentWindowId() === windowId) {
            target = this.parentWindow;
        }
        else {
            target = this.getWindow(windowId);
        }
        if (target) {
            const messagePacket = {
                id: UUID.UUID(),
                originWindowName: window.name,
                date: new Date(),
                message: message
            };
            target.messageQueue.push(messagePacket);
            target.window.postMessage(messagePacket, '*');
        }
    }
    receiveMessage(packet) {
        if (!packet || !packet.message) {
            return;
        } // Ignore other window messages.
        let sendAcknowledge = false;
        switch (packet.message.type) {
            case MessageType.READY: {
                const targetWindow = this.children.find(child => child.windowName === packet.originWindowName);
                // If messages were sent before the window was ready, resend them again.
                if (targetWindow && targetWindow.messageQueue.length > 0) {
                    for (let messagePacket of targetWindow.messageQueue) {
                        messagePacket.date = new Date();
                        targetWindow.window.postMessage(messagePacket, '*');
                    }
                }
                sendAcknowledge = true;
                break;
            }
            case MessageType.WINDOW_CLOSING: {
                const closingWindowIndex = this.children.findIndex(child => child.windowName === packet.originWindowName);
                if (closingWindowIndex > -1) {
                    this.children.splice(closingWindowIndex, 1);
                }
                console.log('Child is closing');
                break;
            }
            case MessageType.ACTION: {
                sendAcknowledge = true;
                this.stream.next(packet.message);
                break;
            }
            case MessageType.ACKNOWLEDGE: {
                let targetWindow;
                if (packet.originWindowName === MAPTOOL_WINDOW_NAME) {
                    targetWindow = this.parentWindow;
                }
                else {
                    targetWindow = this.children.find(child => child.windowName === packet.originWindowName);
                }
                if (targetWindow) {
                    const messageIndex = targetWindow.messageQueue.findIndex(message => message.id === packet.id);
                    if (messageIndex > -1) {
                        targetWindow.messageQueue.splice(messageIndex, 1);
                    }
                }
                break;
            }
            default: {
                console.error(`Received message of unknown type ${packet.message.type}: ${JSON.stringify(packet)}`);
            }
        }
        if (sendAcknowledge) {
            this.broadcastAcknowledge(packet);
        }
    }
    getWindow(id) {
        return this.children.find(child => child.id === id);
    }
    getWindowFeaturesSettings(windowConfig) {
        const left = 0;
        const top = 0;
        const { innerWidth, innerHeight, outerWidth, outerHeight } = window;
        const { availWidth: availScreenWidth, availHeight: availScreenHeight } = window.screen;
        const width = availScreenWidth;
        const height = availScreenHeight - (outerHeight - innerHeight);
        const positionSettings = windowConfig.fullscreen ? `,left=${left},top=${top},width=${width},height=${height}` : '';
        return `resizable=yes,scrollbars=yes,status=yes,location=no${positionSettings}`;
    }
}
WindowMessagingService.ɵprov = i0.ɵɵdefineInjectable({ factory: function WindowMessagingService_Factory() { return new WindowMessagingService(); }, token: WindowMessagingService, providedIn: "root" });
WindowMessagingService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
WindowMessagingService.ctorParameters = () => [];
