// @wf1/core-ui/esm2015/lib/wfml-mapping-container/wfml-mapping-container.component.js
import { Component, Input, TemplateRef } from "@angular/core";
import { TokenService } from "../services/token.service";
import { AppConfigService } from "../services/app-config.service";
export class WFMLMappingContainerComponent {
    //   isMapInitialized = false;
    constructor(token, appConfig) {
        this.token = token;
        this.appConfig = appConfig;
        this.showToolbar = true;
        this.opened = true;
        //   mapControlsOpened = false;
        //   layerTreeSelected = false;
        //   basemapsSelected = false;
        this.isInitializing = true;
        this.isInitializing = token.getOauthToken() === null;
    }
    ngOnInit() {
        if (!this.appConfig.getConfig().application.lazyAuthenticate) {
            this.token.authTokenEmitter.subscribe(() => {
                this.isInitializing = false;
            });
        }
        else {
            this.isInitializing = false;
        }
    }
    ngAfterViewChecked() {
        // if (!this.isInitializing && !this.isMapInitialized) {
        //   this.initializeMap();
        //   this.isMapInitialized = true;
        // }
    }
    toggleSideNav() {
        // this.mapService.onResize();
    }
    initializeMap() {
        // this.mapService.initializeMap('map');
    }
    onMapControlToggled(currentMapControl) {
        // this.mapControlsOpened = currentMapControl !== MapControl.NONE;
        // this.layerTreeSelected = currentMapControl === MapControl.LAYER_TREE;
        // this.basemapsSelected = currentMapControl === MapControl.BASEMAPS;
    }
}
WFMLMappingContainerComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf1-wfml-mapping-container',
                template: "<!--\n  TEMPLATE DECLARATIONS FOR SUB-COMPONENTS\n-->\n<ng-template #wfmlMapPanel>\n  <div id=\"map\"></div>\n</ng-template>\n\n<ng-template #wfmlSidebar>\n  <header>\n    <h4>WILDFIRE INCIDENT MANAGEMENT</h4>\n  </header>\n  <mat-accordion>\n    <mat-expansion-panel>\n      <mat-expansion-panel-header>\n        <mat-panel-title>\n          <i class=\"material-icons\">layers</i>\n        </mat-panel-title>\n        <mat-panel-description>\n          Layers\n        </mat-panel-description>\n      </mat-expansion-panel-header>\n\n      <wf1-wfml-layer-tree></wf1-wfml-layer-tree>\n\n    </mat-expansion-panel>\n    <mat-expansion-panel>\n      <mat-expansion-panel-header>\n        <mat-panel-title>\n          <i class=\"material-icons\">apps</i>\n        </mat-panel-title>\n        <mat-panel-description>\n          Basemaps\n        </mat-panel-description>\n      </mat-expansion-panel-header>\n\n      <mat-button-toggle-group name=\"basemaps\" aria-label=\"Basemaps\">\n        <mat-button-toggle value=\"imagery\">Imagery</mat-button-toggle>\n        <mat-button-toggle value=\"street\">Street</mat-button-toggle>\n        <mat-button-toggle value=\"topographic\">Topographic</mat-button-toggle>\n      </mat-button-toggle-group>\n\n    </mat-expansion-panel>\n  </mat-accordion>\n</ng-template>\n\n<!--\n  MAIN TEMPLATE DECLARATION\n-->\n\n\n<!--\n  MAIN TEMPLATE DECLARATION\n-->\n<div id=\"core-interface-container\">\n  <wf1-application-header [titleText]=\"titleText\"></wf1-application-header>\n\n  <div class=\"wf1-loading-screen\" *ngIf=\"isInitializing\">\n    <h1>Initializing Application</h1>\n    <wf1-spinner></wf1-spinner>\n    <h3>Please wait while the application initializes.</h3>\n  </div>\n\n  <ng-container *ngIf=\"!isInitializing\">\n    <wf1-application-toolbar \n        *ngIf=\"showToolbar\"\n        [menuTemplate]=\"menuTemplate\" \n        [toolbarTemplate]=\"toolbarTemplate\" \n        (selectedMapControl)=\"onMapControlToggled($event)\"\n    ></wf1-application-toolbar>\n\n    <mat-sidenav-container class=\"mat-typography\" hasBackdrop=false>\n      <mat-sidenav id=\"side-nav\" #sidebar mode=\"side\" [(opened)]=\"opened\" >\n        <ng-container *ngTemplateOutlet=\"sidebarTemplate ? sidebarTemplate : wfmlSidebar\"></ng-container>\n      </mat-sidenav>\n      <!-- <mat-sidenav id=\"layers-sidebar\" #mapControls mode=\"over\" hasBackdrop=\"false\" position=\"end\" [(opened)]=\"mapControlsOpened\" > -->\n        <!-- <ng-container *ngIf=\"layerTreeSelected\">\n          <wf1-wfml-layer-tree></wf1-wfml-layer-tree>\n        </ng-container>\n        <ng-container *ngIf=\"basemapsSelected\">\n          <wf1-wfml-basemap-selector></wf1-wfml-basemap-selector>\n        </ng-container> -->\n      <!-- </mat-sidenav> -->\n      <mat-sidenav-content>\n        <div id=\"side-nav-toggle\" (click)=\"sidebar.toggle();toggleSideNav()\" (keydown)=\"sidebar.toggle();toggleSideNav()\">\n          <mat-icon>{{ (opened) ? 'arrow_left' : 'arrow_right' }}</mat-icon>\n        </div>\n        <ng-container *ngTemplateOutlet=\"mainContentTemplate ? mainContentTemplate : wfmlMapPanel\"></ng-container>\n      </mat-sidenav-content>\n    </mat-sidenav-container>\n  </ng-container>\n</div>\n"
            },] }
];
WFMLMappingContainerComponent.ctorParameters = () => [
    { type: TokenService },
    { type: AppConfigService }
];
WFMLMappingContainerComponent.propDecorators = {
    titleText: [{ type: Input }],
    menuTemplate: [{ type: Input }],
    toolbarTemplate: [{ type: Input }],
    sidebarTemplate: [{ type: Input }],
    mainContentTemplate: [{ type: Input }],
    showToolbar: [{ type: Input }]
};
