// @wf1/core-ui/esm2015/lib/info-incident-panel/info-incident-panel.component.js
import { Component, Input } from "@angular/core";
// External
import { select, Store } from "@ngrx/store";
// Redux
import { MapService } from "../services/map.service";
export class InfoIncidentPanelComponent {
    constructor(store, mapService) {
        this.store = store;
        this.mapService = mapService;
        this.codeLabelLookupTables = {};
    }
    ngOnInit() {
        this.getCodeLookupTables();
    }
    getCodeLookupTables() {
        this.store.pipe(select('codeData', 'codeIndex')).subscribe(codeIndex => this.codeLabelLookupTables = codeIndex);
    }
    getLabel(table, value) {
        return this.codeLabelLookupTables[table] && value ? this.codeLabelLookupTables[table][value] : '';
    }
    getPanelClasses() {
        let incidentClass;
        if (this.incident.incidentSituation) {
            incidentClass = (this.incident.incidentSituation.interfaceFireInd)
                ? 'incident-interface'
                : 'incident-non-interface';
        }
        else {
            incidentClass = 'incident-default';
        }
        return `info-panel ${incidentClass}`;
    }
    getFormattedLocation() {
        return (this.incident && this.incident.incidentLocation)
            ? this.mapService.formatCoordinates([this.incident.incidentLocation.longitude, this.incident.incidentLocation.latitude])
            : '';
    }
}
InfoIncidentPanelComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf1-info-incident-panel',
                template: "<div [ngClass]=\"getPanelClasses()\">\n  <h3 class=\"incident-detail-id\">\n    {{incident.incidentLabel}}\n    <span class=\"incident-detail-id-type\">{{getLabel(\"INCIDENT_TYPE_CODE\", incident.incidentTypeCode)}}</span>\n  </h3>\n  <div><strong>Point:</strong> {{getFormattedLocation()}}</div>\n  <div><strong>Incident Commander:</strong> {{incident.incidentCommanderName}}</div>\n  <div><strong>Stage of Control:</strong> {{getLabel(\"STAGE_OF_CONTROL_CODE\", incident.incidentSituation.stageOfControlCode)}}</div>\n  <div><strong>Size:</strong> {{incident.incidentSituation.fireSizeHectares}}</div>\n</div>\n",
                styles: [".info-panel{padding:1rem}.info-panel .incident-detail-id{font-weight:700;padding-bottom:.5rem;margin-bottom:0}.info-panel .incident-detail-id-type{font-weight:400;font-style:italic;margin-left:1rem}.incident-default{color:#fff;background-color:#263238}.incident-interface{color:#fff;background-color:#ff4545}.incident-non-interface{color:#fff;background-color:#009e73}"]
            },] }
];
InfoIncidentPanelComponent.ctorParameters = () => [
    { type: Store },
    { type: MapService }
];
InfoIncidentPanelComponent.propDecorators = {
    incident: [{ type: Input }]
};
