// @busacca/ng-pick-datetime/esm2015/lib/date-time/adapter/moment-adapter/moment-date-time.module.js
/**
 * moment-date-time.module
 */
import { NgModule } from '@angular/core';
import { MomentDateTimeAdapter, OWL_MOMENT_DATE_TIME_ADAPTER_OPTIONS } from './moment-date-time-adapter.class';
import { OWL_MOMENT_DATE_TIME_FORMATS } from './moment-date-time-format.class';
import { DateTimeAdapter, OWL_DATE_TIME_LOCALE } from '../date-time-adapter.class';
import { OWL_DATE_TIME_FORMATS } from '../date-time-format.class';
// import { DateTimeAdapter, OWL_DATE_TIME_FORMATS, OWL_DATE_TIME_LOCALE_PROVIDER } from 'ng-pick-datetime';
export class MomentDateTimeModule {
}
MomentDateTimeModule.decorators = [
    { type: NgModule, args: [{
                providers: [
                    {
                        provide: DateTimeAdapter,
                        useClass: MomentDateTimeAdapter,
                        deps: [OWL_DATE_TIME_LOCALE, OWL_MOMENT_DATE_TIME_ADAPTER_OPTIONS]
                    },
                ],
            },] }
];
const ɵ0 = OWL_MOMENT_DATE_TIME_FORMATS;
export class OwlMomentDateTimeModule {
}
OwlMomentDateTimeModule.decorators = [
    { type: NgModule, args: [{
                imports: [MomentDateTimeModule],
                providers: [{ provide: OWL_DATE_TIME_FORMATS, useValue: ɵ0 }],
            },] }
];
export { ɵ0 };
