// @wf1/orgunit-rest-api/esm5/api/api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
export { CodeHierarchyEndpointsService } from './codeHierarchyEndpoints.service';
import { CodeHierarchyEndpointsService } from './codeHierarchyEndpoints.service';
export { CodeHierarchyListEndpointsService } from './codeHierarchyListEndpoints.service';
import { CodeHierarchyListEndpointsService } from './codeHierarchyListEndpoints.service';
export { CodeTableEndpointsService } from './codeTableEndpoints.service';
import { CodeTableEndpointsService } from './codeTableEndpoints.service';
export { CodeTableListEndpointsService } from './codeTableListEndpoints.service';
import { CodeTableListEndpointsService } from './codeTableListEndpoints.service';
export { LocationAddressService } from './locationAddress.service';
import { LocationAddressService } from './locationAddress.service';
export { LocationContactService } from './locationContact.service';
import { LocationContactService } from './locationContact.service';
export { OrgUnitService } from './orgUnit.service';
import { OrgUnitService } from './orgUnit.service';
export { OrgUnitAreaService } from './orgUnitArea.service';
import { OrgUnitAreaService } from './orgUnitArea.service';
export { OrgUnitAreaListService } from './orgUnitAreaList.service';
import { OrgUnitAreaListService } from './orgUnitAreaList.service';
export { OrgUnitListService } from './orgUnitList.service';
import { OrgUnitListService } from './orgUnitList.service';
export { OrgUnitLocationService } from './orgUnitLocation.service';
import { OrgUnitLocationService } from './orgUnitLocation.service';
export { OrgUnitLocationListService } from './orgUnitLocationList.service';
import { OrgUnitLocationListService } from './orgUnitLocationList.service';
export { OrgUnitTypeConfigService } from './orgUnitTypeConfig.service';
import { OrgUnitTypeConfigService } from './orgUnitTypeConfig.service';
export { TopLevelService } from './topLevel.service';
import { TopLevelService } from './topLevel.service';
/** @type {?} */
export var APIS = [CodeHierarchyEndpointsService, CodeHierarchyListEndpointsService, CodeTableEndpointsService, CodeTableListEndpointsService, LocationAddressService, LocationContactService, OrgUnitService, OrgUnitAreaService, OrgUnitAreaListService, OrgUnitListService, OrgUnitLocationService, OrgUnitLocationListService, OrgUnitTypeConfigService, TopLevelService];
