// @wf1/orgunit-rest-api/esm5/api/orgUnitList.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Wildfire Org Unit REST API
 * The Wildfire Org Unit REST API.
 *
 * OpenAPI spec version: 1.0.0-SNAPSHOT
 *
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */
/* tslint:disable:no-unused-variable member-ordering */
import * as tslib_1 from "tslib";
import { Inject, Injectable, Optional } from '@angular/core';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { CustomHttpUrlEncodingCodec } from '../encoder';
import { BASE_PATH } from '../variables';
import { OrgUnitConfiguration } from '../configuration';
var OrgUnitListService = /** @class */ (function () {
    function OrgUnitListService(httpClient, basePath, configuration) {
        this.httpClient = httpClient;
        this.basePath = 'https://https://d1api.vividsolutions.com/wforg-org-unit-api';
        this.defaultHeaders = new HttpHeaders();
        this.configuration = new OrgUnitConfiguration();
        if (basePath) {
            this.basePath = basePath;
        }
        if (configuration) {
            this.configuration = configuration;
            this.basePath = basePath || configuration.basePath || this.basePath;
        }
    }
    /**
     * @param {?} consumes string[] mime-types
     * @return {?} true: consumes contains 'multipart/form-data', false: otherwise
     */
    OrgUnitListService.prototype.canConsumeForm = /**
     * @param {?} consumes string[] mime-types
     * @return {?} true: consumes contains 'multipart/form-data', false: otherwise
     */
    function (consumes) {
        /** @type {?} */
        var form = 'multipart/form-data';
        try {
            for (var consumes_1 = tslib_1.__values(consumes), consumes_1_1 = consumes_1.next(); !consumes_1_1.done; consumes_1_1 = consumes_1.next()) {
                var consume = consumes_1_1.value;
                if (form === consume) {
                    return true;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (consumes_1_1 && !consumes_1_1.done && (_a = consumes_1.return)) _a.call(consumes_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return false;
        var e_1, _a;
    };
    /**
     * @param {?=} effectiveAsOfDate
     * @param {?=} orgUnitName
     * @param {?=} integerAlias
     * @param {?=} characterAlias
     * @param {?=} orgUnitTypeCode
     * @param {?=} orgUnitAreaTypeCode
     * @param {?=} point
     * @param {?=} pageNumber
     * @param {?=} pageRowCount
     * @param {?=} orderBy
     * @param {?=} restVersion
     * @param {?=} observe
     * @param {?=} reportProgress
     * @return {?}
     */
    OrgUnitListService.prototype.getOrgUnitList1 = /**
     * @param {?=} effectiveAsOfDate
     * @param {?=} orgUnitName
     * @param {?=} integerAlias
     * @param {?=} characterAlias
     * @param {?=} orgUnitTypeCode
     * @param {?=} orgUnitAreaTypeCode
     * @param {?=} point
     * @param {?=} pageNumber
     * @param {?=} pageRowCount
     * @param {?=} orderBy
     * @param {?=} restVersion
     * @param {?=} observe
     * @param {?=} reportProgress
     * @return {?}
     */
    function (effectiveAsOfDate, orgUnitName, integerAlias, characterAlias, orgUnitTypeCode, orgUnitAreaTypeCode, point, pageNumber, pageRowCount, orderBy, restVersion, observe, reportProgress) {
        if (observe === void 0) { observe = 'body'; }
        if (reportProgress === void 0) { reportProgress = false; }
        /** @type {?} */
        var queryParameters = new HttpParams({ encoder: new CustomHttpUrlEncodingCodec() });
        if (effectiveAsOfDate !== undefined) {
            queryParameters = queryParameters.set('effectiveAsOfDate', /** @type {?} */ (effectiveAsOfDate));
        }
        if (orgUnitName !== undefined) {
            queryParameters = queryParameters.set('orgUnitName', /** @type {?} */ (orgUnitName));
        }
        if (integerAlias !== undefined) {
            queryParameters = queryParameters.set('integerAlias', /** @type {?} */ (integerAlias));
        }
        if (characterAlias !== undefined) {
            queryParameters = queryParameters.set('characterAlias', /** @type {?} */ (characterAlias));
        }
        if (orgUnitTypeCode !== undefined) {
            queryParameters = queryParameters.set('orgUnitTypeCode', /** @type {?} */ (orgUnitTypeCode));
        }
        if (orgUnitAreaTypeCode !== undefined) {
            queryParameters = queryParameters.set('orgUnitAreaTypeCode', /** @type {?} */ (orgUnitAreaTypeCode));
        }
        if (point !== undefined) {
            queryParameters = queryParameters.set('point', /** @type {?} */ (point));
        }
        if (pageNumber !== undefined) {
            queryParameters = queryParameters.set('pageNumber', /** @type {?} */ (pageNumber));
        }
        if (pageRowCount !== undefined) {
            queryParameters = queryParameters.set('pageRowCount', /** @type {?} */ (pageRowCount));
        }
        if (orderBy !== undefined) {
            queryParameters = queryParameters.set('orderBy', /** @type {?} */ (orderBy));
        }
        /** @type {?} */
        var headers = this.defaultHeaders;
        if (restVersion !== undefined && restVersion !== null) {
            headers = headers.set('Rest-Version', String(restVersion));
        }
        /** @type {?} */
        var httpHeaderAccepts = [
            'application/json',
            'application/xml'
        ];
        /** @type {?} */
        var httpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
        if (httpHeaderAcceptSelected != undefined) {
            headers = headers.set("Accept", httpHeaderAcceptSelected);
        }
        /** @type {?} */
        var consumes = [];
        return this.httpClient.get(this.basePath + "/orgUnits", {
            params: queryParameters,
            withCredentials: this.configuration.withCredentials,
            headers: headers,
            observe: observe,
            reportProgress: reportProgress
        });
    };
    OrgUnitListService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    OrgUnitListService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: String, decorators: [{ type: Optional }, { type: Inject, args: [BASE_PATH,] }] },
        { type: OrgUnitConfiguration, decorators: [{ type: Optional }] }
    ]; };
    return OrgUnitListService;
}());
export { OrgUnitListService };
if (false) {
    /** @type {?} */
    OrgUnitListService.prototype.basePath;
    /** @type {?} */
    OrgUnitListService.prototype.defaultHeaders;
    /** @type {?} */
    OrgUnitListService.prototype.configuration;
    /** @type {?} */
    OrgUnitListService.prototype.httpClient;
}
