// @wf1/orgunit-rest-api/esm5/api.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule, SkipSelf, Optional } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { OrgUnitConfiguration } from './configuration';
import { CodeHierarchyEndpointsService } from './api/codeHierarchyEndpoints.service';
import { CodeHierarchyListEndpointsService } from './api/codeHierarchyListEndpoints.service';
import { CodeTableEndpointsService } from './api/codeTableEndpoints.service';
import { CodeTableListEndpointsService } from './api/codeTableListEndpoints.service';
import { LocationAddressService } from './api/locationAddress.service';
import { LocationContactService } from './api/locationContact.service';
import { OrgUnitService } from './api/orgUnit.service';
import { OrgUnitAreaService } from './api/orgUnitArea.service';
import { OrgUnitAreaListService } from './api/orgUnitAreaList.service';
import { OrgUnitListService } from './api/orgUnitList.service';
import { OrgUnitLocationService } from './api/orgUnitLocation.service';
import { OrgUnitLocationListService } from './api/orgUnitLocationList.service';
import { OrgUnitTypeConfigService } from './api/orgUnitTypeConfig.service';
import { TopLevelService } from './api/topLevel.service';
var ApiModule = /** @class */ (function () {
    function ApiModule(parentModule) {
        if (parentModule) {
            throw new Error('ApiModule is already loaded. Import your base AppModule only.');
        }
    }
    /**
     * @param {?} configurationFactory
     * @return {?}
     */
    ApiModule.forRoot = /**
     * @param {?} configurationFactory
     * @return {?}
     */
    function (configurationFactory) {
        return {
            ngModule: ApiModule,
            providers: [{ provide: OrgUnitConfiguration, useFactory: configurationFactory }]
        };
    };
    ApiModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, HttpClientModule],
                    declarations: [],
                    exports: [],
                    providers: [
                        CodeHierarchyEndpointsService,
                        CodeHierarchyListEndpointsService,
                        CodeTableEndpointsService,
                        CodeTableListEndpointsService,
                        LocationAddressService,
                        LocationContactService,
                        OrgUnitService,
                        OrgUnitAreaService,
                        OrgUnitAreaListService,
                        OrgUnitListService,
                        OrgUnitLocationService,
                        OrgUnitLocationListService,
                        OrgUnitTypeConfigService,
                        TopLevelService
                    ]
                },] },
    ];
    /** @nocollapse */
    ApiModule.ctorParameters = function () { return [
        { type: ApiModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
    ]; };
    return ApiModule;
}());
export { ApiModule };
