// @wf1/core-ui/esm2015/lib/filter-sidebar-panel/filter-sidebar-panel.component.js
import { Component, Input } from "@angular/core";
import { BasicSidebarPanelComponent } from "../basic-sidebar-panel/basic-sidebar-panel.component";
export class FilterSidebarPanelComponent extends BasicSidebarPanelComponent {
    constructor() {
        super(...arguments);
        this.recordCount = 0;
        this.isInitializing = true;
        this.isShowingFilterOptions = false;
    }
    //   constructor(public token: TokenService) {
    //     super(token);
    //   }
    getMainContentClasses() {
        return (this.isShowingFilterOptions) ? 'side-panel-with-toolbar-and-filter-and-filter-content' : 'side-panel-with-toolbar-and-filter-content';
    }
    setFilterSize(isShowingFilterOptions) {
        this.isShowingFilterOptions = isShowingFilterOptions;
    }
}
FilterSidebarPanelComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf1-filter-sidebar-panel',
                template: "<div class=\"side-panel-container\">\n    <div class=\"side-panel-title-section\">\n        <div class=\"side-panel-link-button side-panel-link-button-left\" *ngIf=\"showBackButton\">\n            <button mat-button (click)=\"titleBackClicked()\" matTooltip=\"Go Back\" [matTooltipShowDelay]=\"TOOLTIP_DELAY\"><mat-icon>arrow_back</mat-icon></button>\n        </div>\n\n        <h3 class=\"side-panel-title\">{{titleText}}</h3>\n        \n        <div [ngClass]=\"getTitleIconClasses()\" *ngIf=\"linkIcon !== undefined\">\n            <button mat-button (click)=\"titleLinkClicked()\" [matTooltip]=\"linkTitle\" [matTooltipShowDelay]=\"TOOLTIP_DELAY\"><mat-icon>{{linkIcon}}</mat-icon></button>\n        </div>\n    </div>\n    <wf1-search-bar [count]=\"recordCount\" [componentId]=\"componentId\">\n        <wf1-sort-modal *ngIf=\"sortOptions && defaultSort\" [options]=\"sortOptions\" [defaultSort]=\"defaultSort\" [componentId]=\"componentId\"></wf1-sort-modal>\n    </wf1-search-bar>\n    <wf1-filters [filters]=\"filters\" (isShowingFilterOptions)=\"setFilterSize($event)\" [componentId]=\"componentId\"></wf1-filters>\n    <div [ngClass]=\"getMainContentClasses()\">\n        <div class=\"interior-panel\">\n            <ng-content></ng-content>\n        </div>\n    </div>\n</div>\n"
            },] }
];
FilterSidebarPanelComponent.propDecorators = {
    recordCount: [{ type: Input }],
    filters: [{ type: Input }],
    defaultSort: [{ type: Input }],
    sortOptions: [{ type: Input }],
    componentId: [{ type: Input }]
};
