// @wf1/core-ui/esm2015/lib/application-toolbar/application-toolbar.component.js
import { Component, EventEmitter, Input, Output, TemplateRef } from "@angular/core";
export var MapControl;
(function (MapControl) {
    MapControl["NONE"] = "none";
    MapControl["LAYER_TREE"] = "layer-tree";
    MapControl["BASEMAPS"] = "basemaps";
})(MapControl || (MapControl = {}));
export class ApplicationToolbarComponent {
    constructor() {
        this.selectedMapControl = new EventEmitter();
        this.isLayersToggled = false;
        this.isBasemapsToggled = false;
    }
    ngOnInit() {
    }
    ngAfterViewInit() {
    }
}
ApplicationToolbarComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf1-application-toolbar',
                template: "<ng-template #defaultToolbar>\n  <button class=\"wf1-menu-button\" mat-button color=\"primary\">\n    <wf1-application-icon iconType=\"incident\" [hasHalo]=\"true\"></wf1-application-icon>\n  </button>\n  <button class=\"wf1-menu-button\" mat-button color=\"primary\">\n    <wf1-application-icon iconType=\"report-of-fire\"></wf1-application-icon>\n  </button>\n</ng-template>\n\n<mat-toolbar class=\"mat-typography wf1-application-toolbar\" color=\"accent\">\n\n  <ng-container *ngIf=\"menuTemplate\">\n    <mat-menu #appMenu=\"matMenu\">\n      <div *ngTemplateOutlet=\"menuTemplate\"></div>\n    </mat-menu>\n\n    <button class=\"wf1-menu-button\" mat-button color=\"primary\" [matMenuTriggerFor]=\"appMenu\" matTooltip=\"Menu\">\n      <mat-icon>menu</mat-icon>\n    </button>\n  </ng-container>\n\n  <div *ngTemplateOutlet=\"toolbarTemplate ? toolbarTemplate: defaultToolbar\"></div>\n\n</mat-toolbar>\n"
            },] }
];
ApplicationToolbarComponent.ctorParameters = () => [];
ApplicationToolbarComponent.propDecorators = {
    menuTemplate: [{ type: Input }],
    toolbarTemplate: [{ type: Input }],
    selectedMapControl: [{ type: Output }]
};
