// @wf1/core-ui/esm2015/lib/basic-sidebar-panel/basic-sidebar-panel.component.js
import { Component, EventEmitter, Input, Output } from "@angular/core";
import { TokenService } from "../services/token.service";
import { NavigationService } from "../services/navigation.service";
export class BasicSidebarPanelComponent {
    constructor(token, navigationService) {
        this.token = token;
        this.navigationService = navigationService;
        this.TOOLTIP_DELAY = 500;
        this.showBackButton = false;
        this.linkClick = new EventEmitter();
        this.isInitializing = true;
        this.isInitializing = token.getOauthToken() === null;
    }
    ngOnInit() {
        this.token.authTokenEmitter.subscribe(() => {
            this.isInitializing = false;
        });
        if (!this.linkTitle) {
            if (this.linkIcon == "launch") {
                this.linkTitle = "Launch Dual Window";
            }
            else if (this.linkIcon == "arrow_back") {
                this.linkTitle = "Back";
            }
            else {
                this.linkTitle = null;
            }
        }
    }
    getTitleIconClasses() {
        return this.linkAlignLeft ? 'side-panel-link-button side-panel-link-button-left' : 'side-panel-link-button side-panel-link-button-right';
    }
    getMainContentClasses() {
        return 'side-panel-content';
    }
    titleLinkClicked() {
        this.linkClick.emit(this.titleText);
    }
    titleBackClicked() {
        this.navigationService.back();
    }
}
BasicSidebarPanelComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf1-basic-sidebar-panel',
                template: "<div class=\"side-panel-container\">\n    <div class=\"side-panel-title-section\">\n        <div class=\"side-panel-link-button side-panel-link-button-left\" *ngIf=\"showBackButton\">\n            <button mat-button (click)=\"titleBackClicked()\" matTooltip=\"Go Back\" [matTooltipShowDelay]=\"TOOLTIP_DELAY\"><mat-icon>arrow_back</mat-icon></button>\n        </div>\n\n        <h3 class=\"side-panel-title\">{{titleText}}</h3>\n\n        <div [ngClass]=\"getTitleIconClasses()\" *ngIf=\"linkIcon !== undefined\">\n            <button mat-button (click)=\"titleLinkClicked()\" [matTooltip]=\"linkTitle\" [matTooltipShowDelay]=\"TOOLTIP_DELAY\"><mat-icon>{{linkIcon}}</mat-icon></button>\n        </div>\n    </div>\n    <div [ngClass]=\"getMainContentClasses()\">\n        <div class=\"interior-panel\">\n            <ng-content></ng-content>\n        </div>\n    </div>\n</div>\n"
            },] }
];
BasicSidebarPanelComponent.ctorParameters = () => [
    { type: TokenService },
    { type: NavigationService }
];
BasicSidebarPanelComponent.propDecorators = {
    titleText: [{ type: Input }],
    linkIcon: [{ type: Input }],
    linkAlignLeft: [{ type: Input }],
    linkTitle: [{ type: Input }],
    showBackButton: [{ type: Input }],
    linkClick: [{ type: Output }]
};
