// @wf1/core-ui/esm2015/lib/location/location-toolbar.component.js
import { Component, EventEmitter, Input, Output } from "@angular/core";
// Models
import { AppConfigService } from "../services/app-config.service";
import { MapService } from "../services/map.service";
export class LocationToolbarComponent {
    // selectedLocation = new FormControl('', this.geometryValidator(this.mapService));
    constructor(config, mapService) {
        this.config = config;
        this.mapService = mapService;
        this.locationSelect = new EventEmitter();
        this.TOOLTIP_DELAY = 500;
        this.MAP_INSTANCE = 'map';
    }
    set location(location) {
        if (location) {
            // this.selectedLocation.setValue(this.mapService.formatCoordinates(location));
            this.locationSelect.emit(location);
        }
    }
    geometryValidator(mapService) {
        return (control) => {
            return null;
            // if (!control.value) {
            //   return null;
            // }
            // const coordinates = mapService.parseCoordinates(control.value);
            // if (!coordinates) {
            //  return { invalidGeom: 'The geometry is invalid' };
            // } else {
            //   this.mapService.setAnchor(this.MAP_INSTANCE, coordinates);
            //   // const formattedCoordinates = mapService.formatCoordinates(coordinates);
            //   // if (formattedCoordinates !== control.value) {
            //   //   control.patchValue(formattedCoordinates, { emitEvent: false })
            //   // }
            // }
            // return null;
        };
    }
    ngOnInit() {
    }
    ngOnDestroy() {
    }
    zoomToSelectedLocation() {
        // this.mapService.zoomToPlaceNameSearchLocation(this.MAP_INSTANCE);
    }
    setAnchorFromMap() {
        // const location = this.mapService.setAnchorFromMap(this.MAP_INSTANCE);
        // if (location) {
        //     this.selectedLocation.setValue(this.mapService.formatCoordinates(location));
        //     this.locationSelect.emit(location);
        // } else if (this.selectedLocation.value) {
        //     this.locationSelect.emit(this.mapService.parseCoordinates(this.selectedLocation.value));
        // } else {
        //     this.locationSelect.emit(null);
        // }
    }
    clearSelectedLocation() {
        // this.selectedLocation.setValue('');
        // this.locationSelect.emit(null);
        // this.mapService.setAnchor(this.MAP_INSTANCE, null);
    }
}
LocationToolbarComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf1-location-toolbar',
                template: "<div class=\"location-toolbar\">\n    <mat-form-field aria-label=\"Selected Location\" class=\"selected-location\">\n        <!-- <input matInput placeholder=\"Selected location (lat, long)\" [formControl]=\"selectedLocation\">\n        <mat-error *ngIf=\"selectedLocation.hasError('required')\">Coordinates are required</mat-error>\n        <mat-error *ngIf=\"selectedLocation.hasError('invalidGeom')\">Coordinates are invalid</mat-error> -->\n    </mat-form-field>\n    <div class=\"location-toolbar-buttons\">\n        <button class=\"location-section-header-button\" mat-button matTooltip=\"Zoom To Selected Location\" [matTooltipShowDelay]=\"TOOLTIP_DELAY\" aria-label=\"Zoom To Selected Location\" (click)=\"zoomToSelectedLocation()\">\n            <mat-icon>place</mat-icon>\n        </button>\n        <button class=\"location-section-header-button\" mat-button matTooltip=\"Set Location From Map\" [matTooltipShowDelay]=\"TOOLTIP_DELAY\" aria-label=\"Set Location From Map\" (click)=\"setAnchorFromMap()\">\n            <mat-icon>map</mat-icon>\n        </button>\n        <button class=\"location-section-header-button\" mat-button matTooltip=\"Clear Search Fields\" [matTooltipShowDelay]=\"TOOLTIP_DELAY\" aria-label=\"Clear Point ID Search Fields\" (click)=\"clearSelectedLocation()\">\n            <mat-icon>clear</mat-icon>\n        </button>\n    </div>\n</div>\n"
            },] }
];
LocationToolbarComponent.ctorParameters = () => [
    { type: AppConfigService },
    { type: MapService }
];
LocationToolbarComponent.propDecorators = {
    location: [{ type: Input }],
    locationSelect: [{ type: Output }]
};
