// @wf1/core-ui/esm2015/lib/full-width-panel/full-width-panel.component.js
import { Component, EventEmitter, Input, Output } from "@angular/core";
import { TokenService } from "../services/token.service";
export class FullWidthPanelComponent {
    constructor(token) {
        this.token = token;
        this.recordCount = 0;
        this.linkClick = new EventEmitter();
        this.isInitializing = true;
        this.isShowingFilterOptions = false;
        this.isInitializing = token.getOauthToken() === null;
    }
    ngOnInit() {
        this.token.authTokenEmitter.subscribe(() => {
            this.isInitializing = false;
        });
    }
    getMainContentClasses() {
        return (this.isShowingFilterOptions) ? 'side-panel-with-toolbar-and-filter-and-filter-content' : 'side-panel-with-toolbar-and-filter-content';
    }
    setFilterSize(isShowingFilterOptions) {
        this.isShowingFilterOptions = isShowingFilterOptions;
    }
    titleLinkClicked() {
        this.linkClick.emit(this.titleText);
    }
}
FullWidthPanelComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf1-full-width-panel',
                template: "<div class=\"side-panel-container\">\n    <div class=\"side-panel-title-section\">\n        <h3 class=\"side-panel-title\">{{titleText}}</h3>\n        <div class=\"side-panel-link-button\" *ngIf=\"linkIcon !== undefined\">\n            <button mat-button (click)=\"titleLinkClicked()\"><mat-icon>{{linkIcon}}</mat-icon></button>\n        </div>\n    </div>\n    <wf1-search-bar [count]=\"recordCount\"></wf1-search-bar>\n    <wf1-filters [filters]=\"filters\" (isShowingFilterOptions)=\"setFilterSize($event)\"></wf1-filters>\n    <div [ngClass]=\"getMainContentClasses()\">\n        <div class=\"interior-panel\">\n            <ng-content></ng-content>\n        </div>\n    </div>\n</div>\n"
            },] }
];
FullWidthPanelComponent.ctorParameters = () => [
    { type: TokenService }
];
FullWidthPanelComponent.propDecorators = {
    titleText: [{ type: Input }],
    linkIcon: [{ type: Input }],
    recordCount: [{ type: Input }],
    filters: [{ type: Input }],
    linkClick: [{ type: Output }]
};
