// @wf1/core-ui/esm2015/lib/application-header/application-header.component.js
import { Component, Input } from "@angular/core";
import { AppConfigService } from "../services/app-config.service";
import { TokenService } from "../services/token.service";
export class ApplicationHeaderComponent {
    constructor(appConfig, token) {
        this.appConfig = appConfig;
        this.token = token;
    }
    ngOnInit() {
        const config = this.appConfig.getConfig();
        this.appVersion = config.application.version;
        this.acronym = config.application.acronym;
        this.environment = config.application.environment;
        this.buildNumber = config.application.buildNumber ? `.${config.application.buildNumber}` : '';
        this.token.credentialsEmitter.subscribe((credentialDetails) => {
            if (credentialDetails.given_name && credentialDetails.family_name) {
                this.userName = `${credentialDetails.given_name} ${credentialDetails.family_name}`;
            }
            else if (credentialDetails.givenName && credentialDetails.familyName) {
                this.userName = `${credentialDetails.givenName} ${credentialDetails.familyName}`;
            }
        });
    }
}
ApplicationHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf1-application-header',
                template: "<mat-toolbar class=\"mat-typography wf1-application-header\" color=\"primary\">\n  <div class=\"bc-logo-div\">\n    <img src=\"assets/images/logo/bcgov-logo.png\" class=\"img-fluid center-block logo\" alt=\"BC Gov\">\n  </div>\n  <div class=\"app-logo-and-title-div\">\n    <img src=\"assets/images/logo/wildfire-one-logo-text-horizontal-light.png\" class=\"img-fluid center-block logo\" alt=\"Wildfire ONE\">\n    <span class=\"mat-title\" id=\"header-title\">- {{ titleText }}</span>\n  </div>\n  <div class=\"system-and-user-info mat-caption\">\n    <div *ngIf=\"appVersion\">\n      <p class=\"header-caption-label\">SYSTEM: </p>\n      <p class=\"header-caption-text\"> {{ acronym }} {{ appVersion }}{{ buildNumber }} {{ environment }} </p>\n    </div>\n    <div *ngIf=\"userName\">\n      <p class=\"header-caption-label\">USER: </p>\n      <p class=\"header-caption-text\"> {{ userName }} </p>\n    </div>\n  </div>\n</mat-toolbar>\n"
            },] }
];
ApplicationHeaderComponent.ctorParameters = () => [
    { type: AppConfigService },
    { type: TokenService }
];
ApplicationHeaderComponent.propDecorators = {
    titleText: [{ type: Input }]
};
