// @wf1/core-ui/esm2015/lib/location-sidebar-panel/location-sidebar-panel.component.js
import { Component, EventEmitter, Input, Output } from "@angular/core";
import { BasicSidebarPanelComponent } from "../basic-sidebar-panel/basic-sidebar-panel.component";
export class LocationSidebarPanelComponent extends BasicSidebarPanelComponent {
    constructor() {
        super(...arguments);
        this.locationSelect = new EventEmitter();
        this.isInitializing = true;
    }
    set location(location) {
        if (location) {
            this.selectedLocation = location;
        }
    }
    //   constructor(public token: TokenService) {
    //       super(token);
    //   }
    locationSelected(coordinates) {
        this.locationSelect.emit(coordinates);
    }
}
LocationSidebarPanelComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf1-location-sidebar-panel',
                template: "<div class=\"side-panel-container\">\n    <div class=\"side-panel-title-section\">\n        <h3 class=\"side-panel-title\">{{titleText}}</h3>\n        <div [ngClass]=\"getTitleIconClasses()\" *ngIf=\"linkIcon !== undefined\">\n            <button mat-button (click)=\"titleLinkClicked()\" [matTooltip]=\"linkTitle\" [matTooltipShowDelay]=\"TOOLTIP_DELAY\"><mat-icon>{{linkIcon}}</mat-icon></button>\n        </div>\n    </div>\n    <wf1-location-toolbar [location]=\"selectedLocation\" (locationSelect)=\"locationSelected($event)\"></wf1-location-toolbar>\n    <div class=\"side-panel-with-toolbar-content\">\n        <div class=\"interior-panel\">\n            <ng-content></ng-content>\n        </div>\n    </div>\n</div>\n"
            },] }
];
LocationSidebarPanelComponent.propDecorators = {
    location: [{ type: Input }],
    locationSelect: [{ type: Output }]
};
