// @wf1/wfcc-application-ui/esm2020/application/components/base.component.mjs
import { Directive, EventEmitter, Input, Output } from "@angular/core";
import * as i0 from "@angular/core";
export class BaseComponent {
    constructor() {
        this.stateChange = new EventEmitter();
        this._state = {
            menu: null
        };
    }
    get title() { return this.config.title; }
    get environment() { return this.config.environment; }
    get userName() { return this.config.userName; }
    get actingOnBehalfOf() { return this.config.actingOnBehalfOf; }
    get version() { return this.config.version; }
    set state(s) {
        this._state = s;
        this.menuState = s.menu;
    }
    get menuState() { return this._state.menu; }
    set menuState(m) {
        if (m == this.menuState)
            return;
        this._state.menu = m;
        // copying the state object is important to ensure that the change is detected
        this.stateChange.emit({ ...this._state });
    }
}
BaseComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: BaseComponent, deps: [], target: i0.ɵɵFactoryTarget.Directive });
BaseComponent.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.9", type: BaseComponent, inputs: { config: "config", state: "state" }, outputs: { stateChange: "stateChange" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: BaseComponent, decorators: [{
            type: Directive
        }], propDecorators: { config: [{
                type: Input
            }], state: [{
                type: Input
            }], stateChange: [{
                type: Output
            }] } });
