// @wf1/core-ui/esm2015/lib/search/components/sort/sort.component.js
import { Component, Input } from "@angular/core";
// External
import { select, Store } from "@ngrx/store";
import * as SearchActions from "../../store/actions";
import { selectCurrentSearchParam } from "../../../utils";
export class SortComponent {
    constructor(store) {
        this.store = store;
        this.isVisible = false;
        this.toggleSortDirection = (dir) => dir === 'ASC' ? 'DESC' : 'ASC';
        this.isActiveParam = (param) => param === this.activeSortParam;
        this.isActiveDir = (dir) => dir === this.activeSortDir;
        this.showArrow = (param, dir) => this.isActiveParam(param) && this.isActiveDir(dir);
    }
    ngOnInit() {
        this.setVisiblility();
        this.setDefaultSort();
    }
    setVisiblility() {
        this.isVisible = true;
        this.store.pipe(select(selectCurrentSearchParam(this.componentId, 'sortModalVisible'))).subscribe((isVisible) => this.isVisible = isVisible);
    }
    setDefaultSort() {
        if (this.defaultSort) {
            this.store.dispatch(new SearchActions.UpdateSortAction(this.defaultSort.param, this.defaultSort.direction, this.componentId));
        }
        this.store.pipe(select(selectCurrentSearchParam(this.componentId, 'sortParam'))).subscribe((param) => this.activeSortParam = param);
        this.store.pipe(select(selectCurrentSearchParam(this.componentId, 'sortDirection'))).subscribe((dir) => this.activeSortDir = dir);
    }
    updateSort(option) {
        const direction = option.param === this.activeSortParam ? this.toggleSortDirection(this.activeSortDir) : 'DESC';
        this.store.dispatch(new SearchActions.UpdateSortAction(option.param, direction, this.componentId));
    }
    close() {
        this.store.dispatch(new SearchActions.ToggleSortModalAction(this.componentId));
    }
}
SortComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf1-sort-modal',
                template: "<div class=\"sort-popup\" [class.sort-popup--visible]=\"isVisible\">\n\t<div class=\"sort-popup__header\">\n\t\t<span class=\"sort-popup__header__text\">Sort</span>\n\t\t<mat-icon class=\"sort-popup__header__icon\" (click)=\"close()\" (keydown)=\"close()\">close</mat-icon>\n\t</div>\n\t<ng-container *ngFor=\"let option of options\">\n\t\t<div class=\"sort-popup__item\" [ngClass]=\"{'sort-popup__item--active': isActiveParam(option.param)}\" (click)=\"updateSort(option)\" (keydown)=\"updateSort(option)\">\n\t\t\t<span class=\"sort-popup__item__text\">{{option.label}}</span>\n\t\t\t<mat-icon class=\"sort-popup__item__icon\" *ngIf=\"!isActiveParam(option.param) || showArrow(option.param, 'DESC')\">\n\t\t\t\tarrow_drop_down\n\t\t\t</mat-icon>\n\t\t\t<mat-icon class=\"sort-popup__item__icon\" *ngIf=\"showArrow(option.param, 'ASC')\">\n\t\t\t\tarrow_drop_up\n\t\t\t</mat-icon>\n\t\t</div>\n\t</ng-container>\n</div>\n"
            },] }
];
SortComponent.ctorParameters = () => [
    { type: Store }
];
SortComponent.propDecorators = {
    defaultSort: [{ type: Input }],
    options: [{ type: Input }],
    componentId: [{ type: Input }]
};
